<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Changelog • plinkQC</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Changelog"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-news">
      <header><div class="navbar navbar-standard navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">plinkQC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-file"></span>
     
    documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/plinkQC.html">Step-by-step QC</a>
    </li>
    <li>
      <a href="../articles/AncestryCheck.html">Ancestry</a>
    </li>
    <li>
      <a href="../articles/HapMap.html">HapMap III reference</a>
    </li>
    <li>
      <a href="../articles/Genomes1000.html">1000 Genomes reference</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="../news/index.html">
    <span class="fa fa-newspaper-o"></span>
     
    news
  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-code"></span>
     
    functions
  </a>
</li>
<li>
  <a href="https://github.com/meyer-lab-cshl/plinkQC" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
      <h1 data-toc-skip>Changelog <small></small></h1>
      <small>Source: <a href="https://github.com/meyer-lab-cshl/plinkQC/blob/HEAD/NEWS.md" class="external-link"><code>NEWS.md</code></a></small>
    </div>

    <div class="section level2">
<h2 class="page-header" data-toc-text="0.3.4" id="plinkqc-034">plinkQC 0.3.4<small>2021-07-15</small><a class="anchor" aria-label="anchor" href="#plinkqc-034"></a></h2>
<div class="section level3">
<h3 id="major-changes-0-3-4">major changes<a class="anchor" aria-label="anchor" href="#major-changes-0-3-4"></a></h3>
<ul><li>Add flag for removing high LD regions and add option for user provided file, allowing removal of high LD regions for non-human organisms; fix missing flag <a href="https://github.com/meyer-lab-cshl/plinkQC/commit/19bc5697f80c4766d91c24d773337b8bf88a3e6f" class="external-link">19bc569</a> and <a href="https://github.com/meyer-lab-cshl/plinkQC/issues/42" class="external-link">#42</a>
</li>
<li>Fixed failing data.no_failIDs.lmiss file in check_snp_missingness function when failIDs were present <a href="https://github.com/meyer-lab-cshl/plinkQC/commit/86adc39764ed6271e947d12063229be0e0189d12" class="external-link">86adc39</a> and <a href="https://github.com/meyer-lab-cshl/plinkQC/issues/43" class="external-link">#43</a>
</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.3.3" id="plinkqc-033">plinkQC 0.3.3<small>2021-02-08</small><a class="anchor" aria-label="anchor" href="#plinkqc-033"></a></h2>
<div class="section level3">
<h3 id="major-changes-0-3-3">major changes<a class="anchor" aria-label="anchor" href="#major-changes-0-3-3"></a></h3>
<ul><li>Output all IBD estimates in <code>run_check_relatedness</code> not just fail IBD estimates <a href="https://github.com/meyer-lab-cshl/plinkQC/commit/d009da90434a83cb4867881e3772b518d17dc2dd" class="external-link">d009da90</a>
</li>
<li>Add midp modifier for –hwe and –hardy as recommended on <a href="https://www.cog-genomics.org/plink/1.9/filter#hwe" class="external-link">plink hwe documentation</a> and discussed in issue <a href="https://github.com/meyer-lab-cshl/plinkQC/issues/27" class="external-link">#27</a>; <a href="https://github.com/meyer-lab-cshl/plinkQC/commit/43230a6ce5f079b46c85b1a7a526213177db1f1b" class="external-link">43230a6c</a>
</li>
<li>Provide HapMap and 1000Genomes individual and population identifiers as default for ancestry check functions; include optional color mapping for these reference populations; raised in <a href="https://github.com/meyer-lab-cshl/plinkQC/issues/37" class="external-link">#37</a>, incorporated in <a href="https://github.com/meyer-lab-cshl/plinkQC/commit/65cfcbaf5b0af3b1d3d4b710a8710331a39e0f6e" class="external-link">65cfcba</a>
</li>
<li>for all individual QC functions return data.frame with FID, IID and measure used for filtering <a href="https://github.com/meyer-lab-cshl/plinkQC/issues/5" class="external-link">#5</a> incorporated in <a href="https://github.com/meyer-lab-cshl/plinkQC/commit/295505bc2d2a6278539e1d078c338bcf3a128338" class="external-link">295505b</a>
</li>
<li>allow for user-supplied individual and marker filtering in perIndividualQC, perMarkerQC and cleanData <a href="https://github.com/meyer-lab-cshl/plinkQC/issues/33" class="external-link">#33</a>
</li>
<li>individual highlighting in perIndividualQC plots; several highlight modalities (text/label, shape, color) <a href="https://github.com/meyer-lab-cshl/plinkQC/issues/38" class="external-link">#38</a>
</li>
</ul></div>
<div class="section level3">
<h3 id="minor-changes-0-3-3">minor changes<a class="anchor" aria-label="anchor" href="#minor-changes-0-3-3"></a></h3>
<ul><li>Add checks in <code>cleanData</code> that can distinguish between miss-specified filters and non-failing samples <a href="https://github.com/meyer-lab-cshl/plinkQC/commit/7432b1ee880ad14c7a078f3a0379f7fdcff1684b" class="external-link">7432b1ee</a>; addresses issue <a href="https://github.com/meyer-lab-cshl/plinkQC/issues/31" class="external-link">#31</a>
</li>
<li>Add checks and tests in <code>check_snp_missingness</code>, <code>check_maf</code> and <code>check_hwe</code> in cases where all samples pass <code>perIndividualQC</code>. <a href="https://github.com/meyer-lab-cshl/plinkQC/commit/31ed6f653902857a0a78ae723e786dd84d240b75" class="external-link">31ed6f65</a>; related to issue <a href="https://github.com/meyer-lab-cshl/plinkQC/issues/31" class="external-link">#31</a>
</li>
<li>Add checks and tests in <code>evaluate_check_sex</code> and <code>evaluate_check_het_imiss</code> for all passing samples and accordingly, no fail samples label in plot. <a href="https://github.com/meyer-lab-cshl/plinkQC/commit/d071a4341b0d4c40220b42f81c412da4d320b8ec" class="external-link">d071a43</a>
</li>
<li>Update dropbox links and add note to vignette <a href="https://github.com/meyer-lab-cshl/plinkQC/commit/9bae9886dc637569698d883654fcf71486b6634e" class="external-link">9bae988</a>
</li>
<li>additional plotting parameters for <code>evaluate_*</code> functions, controlling text and title sizes (<a href="https://github.com/meyer-lab-cshl/plinkQC/commit/b38a348da0a4dd60860543b2d53c4de1df058a54" class="external-link">b38a348</a>, <a href="https://github.com/meyer-lab-cshl/plinkQC/commit/f25b8d05bc3462228afb26bb1caec8d829bf7fca" class="external-link">f25b8d0</a>)</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.3.2" id="plinkqc-032">plinkQC 0.3.2<small>2020-07-07</small><a class="anchor" aria-label="anchor" href="#plinkqc-032"></a></h2>
<div class="section level3">
<h3 id="minor-changes-0-3-2">minor changes<a class="anchor" aria-label="anchor" href="#minor-changes-0-3-2"></a></h3>
<ul><li>Add checks and tests in <code>evaluate_check_ancestry</code> for missing non-reference samples <a href="https://github.com/meyer-lab-cshl/plinkQC/commit/5c03971b92341b5a8cb26a1d5f2e37d637f5cbd7" class="external-link">5c03971</a>
</li>
<li>Clarified tutorials:
<ul><li>R internal file.copy instead of systems <code>cp</code> command to make run on windows: <a href="https://github.com/meyer-lab-cshl/plinkQC/commit/6c91ef3368f601658967f88bc20db94a28307fce" class="external-link">6c91ef3</a>, fixes <a href="https://github.com/meyer-lab-cshl/plinkQC/issues/26" class="external-link">#26</a>
</li>
<li>Add additional filter for A-&gt;T, C-&gt;G variants as suggested in <a href="https://github.com/meyer-lab-cshl/plinkQC/issues/24" class="external-link">#24</a>; <a href="https://github.com/meyer-lab-cshl/plinkQC/commit/11e03756852215915310a836d2e73dbcedf12bfc" class="external-link">11e0375</a>
</li>
</ul></li>
<li>Fixed plotting issues in PCA plot of ancestry check:
<ul><li>allow for supplying names of European reference population <a href="https://github.com/meyer-lab-cshl/plinkQC/commit/ae09e64ee2eaeb8337b60b24b126a9c049d6315f" class="external-link">ae09e64</a>
</li>
<li>provide argument to specify number of populations per legend row <a href="https://github.com/meyer-lab-cshl/plinkQC/commit/c7fe85d555c256bf8be3843803ff1cc445f7898f" class="external-link">c7fe85d</a>
</li>
<li>Ensure correct ordering of population colors when reference population is not HapMap <a href="https://github.com/meyer-lab-cshl/plinkQC/commit/86275b7c281835681f38ab16672db61c288cfed9" class="external-link">86275b7</a>
</li>
</ul></li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.3.1" id="plinkqc-031">plinkQC 0.3.1<a class="anchor" aria-label="anchor" href="#plinkqc-031"></a></h2>
<div class="section level3">
<h3 id="minor-changes-0-3-1">minor changes<a class="anchor" aria-label="anchor" href="#minor-changes-0-3-1"></a></h3>
<ul><li>Fixed dead links in vignettes (caused by migration of repository): <a href="https://github.com/meyer-lab-cshl/plinkQC/commit/da987d8f225aa6aca0596b9c4f6a2484b102bdb6" class="external-link">da987d8</a>
</li>
<li>Added note about chrY in Hapmap data (vignette): <a href="https://github.com/meyer-lab-cshl/plinkQC/commit/e8afbb9842ed9421461a8114ac0a00f7955cf0c0" class="external-link">e8afbb9</a>
</li>
<li>Added note about recommended use of plink1.9 (vs 2.0): <a href="https://github.com/meyer-lab-cshl/plinkQC/commit/b69d3d71d23e9b161176a635fcb2b5a2b524591f" class="external-link">b69d3d7</a>
</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.3.0" id="plinkqc-030">plinkQC 0.3.0<small>2019-10-19</small><a class="anchor" aria-label="anchor" href="#plinkqc-030"></a></h2>
<div class="section level3">
<h3 id="major-changes-0-3-0">major changes<a class="anchor" aria-label="anchor" href="#major-changes-0-3-0"></a></h3>
<ul><li>Relationship filter can handle more complicated relationship scenarios as observed in plant genotype sets (fixed <a href="https://github.com/meyer-lab-cshl/plinkQC/issues/11" class="external-link">#11</a>)</li>
<li>code moved to the meyer-lab repository: <a href="https://github.com/meyer-lab-cshl/plinkQC" class="external-link uri">https://github.com/meyer-lab-cshl/plinkQC</a>
</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.2.3" id="plinkqc-023">plinkQC 0.2.3<small>2019-09-25</small><a class="anchor" aria-label="anchor" href="#plinkqc-023"></a></h2>
<div class="section level3">
<h3 id="major-changes-0-2-3">major changes<a class="anchor" aria-label="anchor" href="#major-changes-0-2-3"></a></h3>
<ul><li>Enable return of overview plots as ggplot object (fixes <a href="https://github.com/meyer-lab-cshl/plinkQC/issues/6" class="external-link">#6</a> in <a href="https://github.com/meyer-lab-cshl/plinkQC/commit/ab2e840f0f22ccdeb5317475698aa58e7eecd345" class="external-link">ab2e840</a> and <a href="https://github.com/meyer-lab-cshl/plinkQC/commit/101e74e318752b2f8038330245d875f6afd5b57b" class="external-link">101e74e</a>)</li>
<li>Relationship filter now deals with more complicated relationship scenarios as observed in plant genotype sets (addresses <a href="https://github.com/meyer-lab-cshl/plinkQC/issues/11" class="external-link">#11</a>)</li>
</ul></div>
<div class="section level3">
<h3 id="minor-changes-0-2-3">minor changes<a class="anchor" aria-label="anchor" href="#minor-changes-0-2-3"></a></h3>
<ul><li>give user option to choose maf threshold for relatedness filtering (relates to <a href="https://github.com/meyer-lab-cshl/plinkQC/issues/3" class="external-link">#3</a>)</li>
</ul></div>
<div class="section level3">
<h3 id="bug-fixes-0-2-3">bug fixes<a class="anchor" aria-label="anchor" href="#bug-fixes-0-2-3"></a></h3>
<ul><li>Include check for zero related individuals fixing <a href="https://github.com/meyer-lab-cshl/plinkQC/issues/3" class="external-link">#3</a> in <a href="https://github.com/meyer-lab-cshl/plinkQC/commit/1445a88d41d985e73ffdd161144229683f8352cd" class="external-link">1445a88</a>
</li>
<li>Include check in case all samples fail perIndividual QC in <a href="https://github.com/meyer-lab-cshl/plinkQC/commit/894acc1fa03dadfe0ad2028888142171bcc641eb" class="external-link">894acc1</a> and <a href="https://github.com/meyer-lab-cshl/plinkQC/commit/04642246d18ed4eaac5b9d9a6931d1ecb08308e8" class="external-link">0464224</a>)</li>
<li>Include checks for diagonal derived relationship estimates, and estimate data containing only related individuals; addresses <a href="https://github.com/meyer-lab-cshl/plinkQC/issues/11" class="external-link">#11</a>
</li>
<li>Fix command for genotype conversion in 1000Genomes vignette, addressing issue <a href="https://github.com/meyer-lab-cshl/plinkQC/issues/10" class="external-link">#10</a>
</li>
<li>fix missing rownames error for overviewPerIndividualQC, when relatedness check was included (issue <a href="https://github.com/meyer-lab-cshl/plinkQC/issues/16" class="external-link">#16</a>, <a href="https://github.com/meyer-lab-cshl/plinkQC/commit/fc7a38b1f2b345d9c6c5d69f5dcf0bc57a857f62" class="external-link">fc7a38b</a>)</li>
<li>fix vignette mismatch (issue <a href="https://github.com/meyer-lab-cshl/plinkQC/issues/16" class="external-link">#16</a>, <a href="https://github.com/meyer-lab-cshl/plinkQC/commit/09dcd59e77178b35747aae81a5c1988712e20de9" class="external-link">09dcd59</a>)</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.2.2" id="plinkqc-022">plinkQC 0.2.2<small>2019-06-19</small><a class="anchor" aria-label="anchor" href="#plinkqc-022"></a></h2>
<div class="section level3">
<h3 id="minor-changes-0-2-2">minor changes<a class="anchor" aria-label="anchor" href="#minor-changes-0-2-2"></a></h3>
<ul><li>Fix IDs written to fail.IDs file: previous versions wrote IID,IID, now fixed to FID, IID (fixes <a href="https://github.com/meyer-lab-cshl/plinkQC/issues/2" class="external-link">#2</a>).</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.2.1" id="plinkqc-021">plinkQC 0.2.1<small>2019-03-01</small><a class="anchor" aria-label="anchor" href="#plinkqc-021"></a></h2>
<div class="section level3">
<h3 id="minor-changes-0-2-1">minor changes<a class="anchor" aria-label="anchor" href="#minor-changes-0-2-1"></a></h3>
<ul><li>Fix path check bug in checkPlink</li>
<li>Include test data in build!</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.2.0" id="plinkqc-020">plinkQC 0.2.0<small>2018-11-09</small><a class="anchor" aria-label="anchor" href="#plinkqc-020"></a></h2>
<div class="section level3">
<h3 id="major-changes-0-2-0">major changes<a class="anchor" aria-label="anchor" href="#major-changes-0-2-0"></a></h3>
<ul><li>All system calls to plink are conducted with sys::exec_wait - this should solve platform dependent issues, mainly compatibility with windows.</li>
<li>Make path construction compatible with windows.</li>
<li>path2plink now requires full path to plink executable, no tilde expansion or simple pointer to directory supported.</li>
<li>Fix bug in return of cleanData function: list now contains keep and fail IDs.</li>
<li>Fix bug in return of maf computation: if fail.IDs does not exist, set fail_samples to zero.</li>
</ul></div>
<div class="section level3">
<h3 id="minor-changes-0-2-0">minor changes<a class="anchor" aria-label="anchor" href="#minor-changes-0-2-0"></a></h3>
<ul><li>IBD-fail.IDs now saved without column names to be consistent with other xxx-fail.IDs files.</li>
<li>Include additional progress messages in cleanData()</li>
<li>Remove default double-specification of mafTh and macTh</li>
<li>use checkPlink to return correct path2plink and export to make checkPlink directly accesible to user.</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.1.1" id="plinkqc-011">plinkQC 0.1.1<small>2018-11-06</small><a class="anchor" aria-label="anchor" href="#plinkqc-011"></a></h2>
<div class="section level3">
<h3 id="major-changes-0-1-1">major changes<a class="anchor" aria-label="anchor" href="#major-changes-0-1-1"></a></h3>
<ul><li>run_check_relatedness will only save IBD estimates of individuals whose estimates are higher than the threshold.</li>
</ul></div>
<div class="section level3">
<h3 id="minor-changes-0-1-1">minor changes<a class="anchor" aria-label="anchor" href="#minor-changes-0-1-1"></a></h3>
<ul><li>Fix examples in vignettes 1000 Genomes and HapMap III reference.</li>
<li>Change file access in function examples</li>
<li>Add additional checks in check_ancestry and fix missing refSamplesFile test</li>
</ul></div>
</div>
    <div class="section level2">
<h2 class="page-header" data-toc-text="0.1.0" id="plinkqc-010">plinkQC 0.1.0<small>2018-11-04</small><a class="anchor" aria-label="anchor" href="#plinkqc-010"></a></h2>
<ul><li>Added a <code>NEWS.md</code> file to track changes to the package.</li>
</ul></div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>

</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Hannah Meyer.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

