<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Ancestry estimation based on reference samples of known ethnicities • plinkQC</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Ancestry estimation based on reference samples of known ethnicities">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="standard" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">plinkQC</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html" aria-label="ho"><span class="fa fa-home"></span></a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-documentation" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Documentation"><span class="fa fa-file"></span> Documentation</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-documentation">
<li><a class="dropdown-item" href="../articles/plinkQC.html">Step-by-step QC</a></li>
    <li><a class="dropdown-item" href="../articles/AncestryCheck.html">Ancestry</a></li>
    <li><a class="dropdown-item" href="../articles/HapMap.html">HapMap III reference</a></li>
    <li><a class="dropdown-item" href="../articles/Genomes1000.html">1000 Genomes reference</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="nav-link" href="../news/index.html"><span class="fa fa-newspaper-o"></span> news</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html"><span class="fa fa-code"></span> functions</a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/meyer-lab-cshl/plinkQC"><span class="fa fa-github fa-lg"></span> github</a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Ancestry estimation based on reference samples of known ethnicities</h1>
                        <h4 data-toc-skip class="author">Maha Syed and
Caroline Walter</h4>
            
            <h4 data-toc-skip class="date">2025-11-19</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/meyer-lab-cshl/plinkQC/blob/HEAD/vignettes/AncestryCheck.Rmd" class="external-link"><code>vignettes/AncestryCheck.Rmd</code></a></small>
      <div class="d-none name"><code>AncestryCheck.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="training-your-own-random-forest-classifier">Training Your Own Random Forest Classifier<a class="anchor" aria-label="anchor" href="#training-your-own-random-forest-classifier"></a>
</h2>
<p>For non-human genotype data, there might still be the need to detect
and predict population structure. In addition, in some rare cases, there
might not be a large enough overlap in the SNP genotypes and the trained
reference data (we recommend &gt; 5% overlap). In this case, we
recommend training your own random forest classifier. In the following,
we will demonstrate the workflow for setting up this training, starting
with reference data download. As an example, we show the process for the
1000 genomes dataset. The download and pre-processing are described in
detail in <a href="Genomes1000.html">Processing 1000 Genomes reference
data for ancestry estimation</a>. We will use a test study dataset
included in the package as an example for the workflow.</p>
<div class="section level3">
<h3 id="setting-up-bash-script-variables">Setting up Bash Script Variables<a class="anchor" aria-label="anchor" href="#setting-up-bash-script-variables"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="va">qcdir</span><span class="op">=</span><span class="st">'~/qcdir'</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="va">name</span><span class="op">=</span><span class="st">'data.hg38'</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="va">refdir</span><span class="op">=</span>~/reference</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-r</span> <span class="va">$qcdir</span>/plink_log</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="putting-study-data-in-qcdir">Putting study data in qcdir<a class="anchor" aria-label="anchor" href="#putting-study-data-in-qcdir"></a>
</h3>
<p>Copy your study dataset to the qcdir folder. In this example, we are
assuming that the provided test dataset included with the package has
been moved.</p>
</div>
<div class="section level3">
<h3 id="download-reference-data">Download reference data<a class="anchor" aria-label="anchor" href="#download-reference-data"></a>
</h3>
<p>A suitable reference dataset should be downloaded and if necessary,
re-formated into PLINK format. Vignettes <a href="https://meyer-lab-cshl.github.io/plinkQC/articles/HapMap.html">‘Processing
HapMap III reference data for ancestry estimation’</a> and <a href="https://meyer-lab-cshl.github.io/plinkQC/articles/Genomes1000.html">‘Processing
1000Genomes reference data for ancestry estimation’</a>, show the
download and processing of the HapMap phase III and 1000Genomes phase
III dataset, respectively. In this example, we will use the 1000Genomes
data as the reference dataset. We recommend using
<code>check_relatedness</code> on the reference to filter our related
samples. This prevents data leakage between the training and testing
datasets.</p>
</div>
<div class="section level3">
<h3 id="set-up">Set-up<a class="anchor" aria-label="anchor" href="#set-up"></a>
</h3>
<p>We will first set up some bash variables and create directories
needed; storing the names and directories of the reference and study
will make it easy to use updated versions of the reference or new
datasets in the future. We create a directory named ‘qcdir’ for the
study data. In order to keep the data directory tidy, we will create a
directory for the log files and move them to the log directory for
future reference after each analysis step.</p>
<p>Once you have created ‘qcdir’, be sure to copy the data.bim,
data.bed, and data.fam files into your directory.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="va">qcdir</span><span class="op">=</span>~/qcdir</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="va">refdir</span><span class="op">=</span>~/reference</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="va">name</span><span class="op">=</span><span class="st">'data'</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="va">refname</span><span class="op">=</span><span class="st">'all_hg38'</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a><span class="fu">mkdir</span> <span class="at">-p</span> <span class="va">$qcdir</span>/plink_log</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="match-study-genotypes-and-reference-data">Match study genotypes and reference data<a class="anchor" aria-label="anchor" href="#match-study-genotypes-and-reference-data"></a>
</h3>
<p>In order to compute joint principal components of the reference and
study population, we will need to combine the two datasets. The plink
–merge function enables this merge, but requires the variants in the
datasets to be matching by chromosome, position and alleles. The
following sections show how to extract the relevant data from the
reference and study dataset and how to filter matching variants.</p>
</div>
<div class="section level3">
<h3 id="filter-reference-and-study-data-for-non-a-t-or-g-c-snps">Filter reference and study data for non A-T or G-C SNPs<a class="anchor" aria-label="anchor" href="#filter-reference-and-study-data-for-non-a-t-or-g-c-snps"></a>
</h3>
<p>We will use an awk script to find non-A/T and non-G/C SNPs, as these
SNPs are more difficult to align, and remove them from both the
reference and study data sets. In addition, we will only keep the
autosomes (chr 1-22), only keep snps (snps-only), of those, keep only
the biallelic ones (max-alleles 2), and remove any duplicates (rm-dup).
The data should start off in Plink 1.9 data format.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">awk</span> <span class="st">'BEGIN {OFS="\t"}  ($5$6 == "GC" || $5$6 == "CG" \</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="st">                        || $5$6 == "AT" || $5$6 == "TA")  {print $2}'</span> <span class="dt">\</span></span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>    <span class="va">$qcdir</span>/<span class="va">$name</span>.bim  <span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>    <span class="va">$qcdir</span>/<span class="va">$name</span>.ac_gt_snps</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="fu">awk</span> <span class="st">'BEGIN {OFS="\t"}  ($5$6 == "GC" || $5$6 == "CG" \</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="st">                        || $5$6 == "AT" || $5$6 == "TA")  {print $2}'</span> <span class="dt">\</span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>    <span class="va">$refdir</span>/<span class="va">$refname</span>.bim  <span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>    <span class="va">$qcdir</span>/<span class="va">$refname</span>.ac_gt_snps</span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a>   </span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a><span class="ex">plink2</span> <span class="at">--bfile</span>  <span class="va">$refdir</span>/<span class="va">$refname</span> <span class="dt">\</span></span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a>      <span class="at">--rm-dup</span> exclude-all <span class="dt">\</span></span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a>      <span class="at">--max-alleles</span> 2 <span class="dt">\</span></span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a>      <span class="at">--snps-only</span> just-acgt <span class="dt">\</span></span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a>      <span class="at">--exclude</span> <span class="va">$qcdir</span>/<span class="va">$refname</span>.ac_gt_snps <span class="dt">\</span></span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a>      <span class="at">--chr</span> 1-22 <span class="dt">\</span></span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a>      <span class="at">--make-bed</span> <span class="dt">\</span></span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a>      <span class="at">--out</span> <span class="va">$qcdir</span>/<span class="va">$refname</span>.no_ac_gt_snps</span>
<span id="cb3-18"><a href="#cb3-18" tabindex="-1"></a><span class="fu">mv</span>  <span class="va">$qcdir</span>/<span class="va">$refname</span>.no_ac_gt_snps.log <span class="va">$qcdir</span>/plink_log/<span class="va">$refname</span>.no_ac_gt_snps.log</span>
<span id="cb3-19"><a href="#cb3-19" tabindex="-1"></a></span>
<span id="cb3-20"><a href="#cb3-20" tabindex="-1"></a><span class="ex">plink2</span> <span class="at">--bfile</span>  <span class="va">$qcdir</span>/<span class="va">$name</span> <span class="dt">\</span></span>
<span id="cb3-21"><a href="#cb3-21" tabindex="-1"></a>      <span class="at">--rm-dup</span> exclude-all <span class="dt">\</span></span>
<span id="cb3-22"><a href="#cb3-22" tabindex="-1"></a>      <span class="at">--max-alleles</span> 2 <span class="dt">\</span></span>
<span id="cb3-23"><a href="#cb3-23" tabindex="-1"></a>      <span class="at">--snps-only</span> just-acgt <span class="dt">\</span></span>
<span id="cb3-24"><a href="#cb3-24" tabindex="-1"></a>      <span class="at">--exclude</span> <span class="va">$qcdir</span>/<span class="va">$name</span>.ac_gt_snps <span class="dt">\</span></span>
<span id="cb3-25"><a href="#cb3-25" tabindex="-1"></a>      <span class="at">--make-bed</span> <span class="dt">\</span></span>
<span id="cb3-26"><a href="#cb3-26" tabindex="-1"></a>      <span class="at">--chr</span> 1-22 <span class="dt">\</span></span>
<span id="cb3-27"><a href="#cb3-27" tabindex="-1"></a>      <span class="at">--allow-extra-chr</span> <span class="dt">\</span></span>
<span id="cb3-28"><a href="#cb3-28" tabindex="-1"></a>      <span class="at">--out</span> <span class="va">$qcdir</span>/<span class="va">$name</span>.no_ac_gt_snps</span>
<span id="cb3-29"><a href="#cb3-29" tabindex="-1"></a><span class="fu">mv</span>  <span class="va">$qcdir</span>/<span class="va">$name</span>.no_ac_gt_snps.log <span class="va">$qcdir</span>/plink_log/<span class="va">$name</span>.no_ac_gt_snps.log</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="renaming-variant-identifiers">Renaming variant identifiers<a class="anchor" aria-label="anchor" href="#renaming-variant-identifiers"></a>
</h3>
<p>Variant identifiers can vary between studies. To ensure that the data
can be properly processed, we will rename the variant identifiers to a
common scheme based on chromosome number and basepair location by using
<em>plinkQC</em>’s <code>rename_variant_identifiers</code> function.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://meyer-lab-cshl.github.io/plinkQC/">plinkQC</a></span><span class="op">)</span></span>
<span><span class="va">name</span> <span class="op">&lt;-</span> <span class="st">"data.no_ac_gt_snps"</span></span>
<span><span class="va">refname</span> <span class="op">&lt;-</span> <span class="st">"all_hg38.no_ac_gt_snps"</span></span>
<span><span class="va">path2plink2</span> <span class="op">&lt;-</span> <span class="st">"/Users/syed/bin/plink2"</span></span>
<span><span class="fu"><a href="../reference/rename_variant_identifiers.html">rename_variant_identifiers</a></span><span class="op">(</span>indir<span class="op">=</span><span class="va">qcdir</span>, qcdir<span class="op">=</span><span class="va">qcdir</span>, name<span class="op">=</span><span class="va">name</span>, </span>
<span>                           path2plink2 <span class="op">=</span> <span class="va">path2plink2</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="filtering-out-shared-snps-between-study-and-reference-dataset">Filtering out shared SNPs between study and reference dataset<a class="anchor" aria-label="anchor" href="#filtering-out-shared-snps-between-study-and-reference-dataset"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$qcdir</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="fu">awk</span> <span class="st">'{print $3}'</span> <span class="va">$name</span>.renamed.pvar <span class="op">&gt;</span> <span class="va">$name</span>.renamed.varids.txt</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="ex">plink2</span> <span class="at">--pfile</span> <span class="va">$refname</span>.renamed <span class="at">--extract</span> <span class="va">$name</span>.renamed.varids.txt <span class="at">--make-bed</span> <span class="at">--out</span> <span class="va">$refname</span>.renamed.studysnps</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="conducting-markerqc-pruning-ld-and-individual-qc">Conducting markerQC, pruning LD, and individual QC<a class="anchor" aria-label="anchor" href="#conducting-markerqc-pruning-ld-and-individual-qc"></a>
</h3>
<p>After filtering for SNPS shared in the study, we will conduct marker
and sample quality control on the reference dataset. This is to ensure
only robust data is used as a reference. To reduce correlations between
SNPs, we will prune for variants in linkage disequilibrium (LD) after
removing low quality markers. This is done with <em>plinkQC</em>’s
pruning_ld() function. As part of the function, ranges of known high-LD
are excluded. These ranges are originally provided by <span class="citation">[1]</span>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">refname</span> <span class="op">&lt;-</span> <span class="st">"all_hg38.renamed.studysnps"</span></span>
<span></span>
<span><span class="va">fail_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perMarkerQC.html">perMarkerQC</a></span><span class="op">(</span>indir<span class="op">=</span><span class="va">indir</span>, qcdir<span class="op">=</span><span class="va">qcdir</span>, name<span class="op">=</span><span class="va">name</span>,</span>
<span>                            path2plink<span class="op">=</span><span class="va">path2plink</span>,</span>
<span>                            verbose<span class="op">=</span><span class="cn">TRUE</span>, interactive<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                            showPlinkOutput<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">marker_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cleanData.html">cleanData</a></span><span class="op">(</span>indir <span class="op">=</span> <span class="va">indir</span>, qcdir <span class="op">=</span> <span class="va">qcdir</span>, name <span class="op">=</span> <span class="va">name</span>, </span>
<span>                path2plink <span class="op">=</span> <span class="va">path2plink</span>, filterAncestry <span class="op">=</span> <span class="cn">FALSE</span>, filterRelated <span class="op">=</span> <span class="cn">FALSE</span>, macTh <span class="op">=</span> <span class="cn">NULL</span>, mafTh <span class="op">=</span> <span class="fl">0.05</span>, verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                filterSex <span class="op">=</span> <span class="cn">FALSE</span>, filterHeterozygosity <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                filterSNPMissingness <span class="op">=</span> <span class="cn">TRUE</span>, filterSampleMissingness <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                filterMAF <span class="op">=</span> <span class="cn">TRUE</span>, filterHWE <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">name</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">name</span>, <span class="st">".clean"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/pruning_ld.html">pruning_ld</a></span><span class="op">(</span>indir <span class="op">=</span> <span class="va">indir</span>, qcdir <span class="op">=</span> <span class="va">qcdir</span>, name <span class="op">=</span> <span class="va">name</span>, </span>
<span>      path2plink <span class="op">=</span> <span class="va">path2plink</span>, genomebuild<span class="op">=</span><span class="st">"hg38"</span><span class="op">)</span></span>
<span><span class="va">name</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">name</span>, <span class="st">".pruned"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span>
<span></span>
<span><span class="va">fail_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perIndividualQC.html">perIndividualQC</a></span><span class="op">(</span>indir<span class="op">=</span><span class="va">indir</span>, qcdir<span class="op">=</span><span class="va">qcdir</span>, name<span class="op">=</span><span class="va">name</span>, dont.check_sex <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                                dont.check_relatedness <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                            path2plink<span class="op">=</span><span class="va">path2plink</span>, dont.check_ancestry <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                            interactive<span class="op">=</span><span class="cn">TRUE</span>, verbose<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sample_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cleanData.html">cleanData</a></span><span class="op">(</span>indir <span class="op">=</span> <span class="va">indir</span>, qcdir <span class="op">=</span> <span class="va">qcdir</span>, name <span class="op">=</span> <span class="va">name</span>, </span>
<span>             path2plink <span class="op">=</span> <span class="va">path2plink</span>, filterAncestry <span class="op">=</span> <span class="cn">FALSE</span>, filterRelated <span class="op">=</span> <span class="cn">FALSE</span>,verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            filterSex <span class="op">=</span> <span class="cn">FALSE</span>, filterHeterozygosity <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>            filterSNPMissingness <span class="op">=</span> <span class="cn">FALSE</span>, filterSampleMissingness <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>            filterMAF <span class="op">=</span> <span class="cn">FALSE</span>, filterHWE <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="pca">PCA<a class="anchor" aria-label="anchor" href="#pca"></a>
</h3>
<p>After conducting quality control, we will compute the joint principal
components.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="ex">plink2</span> <span class="at">--bfile</span>  <span class="va">$refname</span>.renamed.studysnps.cleaned.pruned.cleaned <span class="dt">\</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>       <span class="at">--nonfounders</span> <span class="dt">\</span></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>       <span class="at">--freq</span> counts <span class="dt">\</span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>       <span class="at">--pca</span> allele-wts 20 <span class="dt">\</span></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>       <span class="at">--out</span> <span class="va">$refname</span>.renamed.studysnps.cleaned.pruned.cleaned.pca</span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a>       </span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a><span class="fu">mv</span> <span class="va">$qcdir</span>/<span class="va">$refname</span>.renamed.studysnps.cleaned.pruned.cleaned.pca <span class="va">$qcdir</span>/plink_log</span></code></pre></div>
<p>After running a PCA, we will project the original dataset onto the
principal components to rescale. This ensures that future projections
will be on the same scale.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="ex">plink2</span> <span class="at">--pfile</span> <span class="va">$refname</span>.renamed.studysnps.cleaned.pruned.cleaned</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>       <span class="ex">--read-freq</span> <span class="va">$refname</span>.renamed.studysnps.cleaned.pruned.cleaned.pca.acount </span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>       <span class="ex">--score</span> <span class="va">$refname</span>.renamed.studysnps.cleaned.pruned.cleaned.pca.eigenvec.allele 2 6 header-read no-mean-imputation variance-standardize </span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>       <span class="ex">--score-col-nums</span> 7-26 <span class="at">--out</span> <span class="va">$refname</span>.renamed.studysnps.cleaned.pruned.cleaned.projection</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="training-a-random-forest-classifier-in-r">Training a random forest classifier in R<a class="anchor" aria-label="anchor" href="#training-a-random-forest-classifier-in-r"></a>
</h3>
<p>The principal components of the reference dataset will be used to
train the random forest algorithm in R.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://www.stat.berkeley.edu/~breiman/RandomForests/" class="external-link">randomForest</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">filepath</span> <span class="op">&lt;-</span> <span class="st">'insert path to sscore file here'</span></span>
<span><span class="va">proj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span></span>
<span>  file<span class="op">=</span> <span class="va">filepath</span>,</span>
<span>  sep<span class="op">=</span><span class="st">'\t'</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">package.dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/find.package.html" class="external-link">find.package</a></span><span class="op">(</span><span class="st">'plinkQC'</span><span class="op">)</span></span>
<span><span class="va">ancestry_info</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_delim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">package.dir</span>,<span class="st">"extdata/Genomes1000_ID2Pop.txt"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">superpop</span> <span class="op">&lt;-</span> </span>
<span>  <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_delim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">package.dir</span>,<span class="st">"extdata/AncestryGeoLocations.csv"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We divide up the data into a 70% training portion and 30% testing
portion. To ensure a balanced model, we will sample the data to ensure
an even representation of each ancestral superpopulation group in the
testing portion of the data. The number of inidividuals from each group
that is chosen will have to adjusted for the size of the reference
dataset. For the 1000 genomes dataset, we use 448 samples in each of the
five ancestral groups for a roughly 70% of the data.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">proj</span> <span class="op">&lt;-</span> <span class="va">proj</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">ALLELE_CT</span>, <span class="va">NAMED_ALLELE_DOSAGE_SUM</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">proj</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"IID"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"PC"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">labeled_proj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">proj</span>, <span class="va">superpop</span><span class="op">)</span></span>
<span></span>
<span><span class="va">n_individuals</span> <span class="op">&lt;-</span> <span class="fl">448</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html" class="external-link">split</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">labeled_proj</span><span class="op">)</span><span class="op">)</span>, <span class="va">labeled_proj</span><span class="op">$</span><span class="va">Ancestry</span><span class="op">)</span></span>
<span></span>
<span><span class="va">train.ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">idx</span>, <span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">i</span>, <span class="va">n_individuals</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ids_to_keeps</span><span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">train.ids</span>, use.names <span class="op">=</span> <span class="cn">FALSE</span>, recursive <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">train_proj_1000g</span> <span class="op">&lt;-</span> <span class="va">labeled_proj</span><span class="op">[</span><span class="va">ids_to_keeps</span>,<span class="op">]</span></span></code></pre></div>
<p>Then, we can train the random forest. We are providing a starting
value for the parameter but we recommend utilizing a grid search
(described in detail more below) to determine the optimal values.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">train_proj_1000g</span><span class="op">$</span><span class="va">Ancestry</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">train_proj_1000g</span><span class="op">$</span><span class="va">Ancestry</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ancestry_rf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/randomForest/man/randomForest.html" class="external-link">randomForest</a></span><span class="op">(</span><span class="va">Ancestry</span> <span class="op">~</span> <span class="va">.</span>,</span>
<span>                        data <span class="op">=</span> <span class="va">train_proj_1000g</span><span class="op">[</span>,<span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">23</span><span class="op">)</span><span class="op">]</span>, </span>
<span>                        method <span class="op">=</span> <span class="st">"rf"</span>, </span>
<span>                        ntree <span class="op">=</span> <span class="fl">750</span>,</span>
<span>                        importance <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">ancestry_rf</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="predicting-ancestries-of-new-study-data">Predicting ancestries of new study data<a class="anchor" aria-label="anchor" href="#predicting-ancestries-of-new-study-data"></a>
</h3>
<p>To predict the ancestries of a new study dataset with our trained
random forest, you need to project the newdata onto the principal
components of the reference dataset. This is done with the –score
function in plink2. The projection is then used as input for the random
forest.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="ex">plink2</span> <span class="at">--pfile</span> newdata </span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a>       <span class="ex">--read-freq</span> all_hg38.pca.acount </span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a>       <span class="ex">--score</span> all_hg38.pca.eigenvec.allele 2 6 header-read no-mean-imputation variance-standardize </span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a>       <span class="ex">--score-col-nums</span> 7-26 <span class="at">--out</span> newdata.projection</span></code></pre></div>
<p>Then, we can use the random forest to predict the ancestry of the new
data.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">filepath</span> <span class="op">&lt;-</span> <span class="st">"Insert path to newdata.sscore here"</span></span>
<span><span class="va">newdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span></span>
<span>  file<span class="op">=</span> <span class="va">filepath</span>, </span>
<span>  sep<span class="op">=</span><span class="st">'\t'</span>, header <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">newdata</span> <span class="op">&lt;-</span> <span class="va">newdata</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">ALLELE_CT</span>, <span class="va">NAMED_ALLELE_DOSAGE_SUM</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">newdata</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ID"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"PC"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">predictions</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">ancestry_rf</span>, <span class="va">newdata</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="evalulating-and-tuning-classification">Evalulating and Tuning Classification<a class="anchor" aria-label="anchor" href="#evalulating-and-tuning-classification"></a>
</h3>
<p>To evaluate the accuracy of the model, a commonly used metric is the
out-of-bag (OOB) error rate and confusion matrix. The confusion matrix
displays the number of correct classifications for each ancestry along
the diagonal. Any values outside the diagonal are misclassifications for
other ancestries. A summary of these metrics can be seen through:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ancestry_rf</span></span></code></pre></div>
<p>And we can visualize the confusion matrix as a heatmap.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_rf</span> <span class="op">&lt;-</span> <span class="va">ancestry_rf</span><span class="op">$</span><span class="va">confusion</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/heatmap.html" class="external-link">heatmap</a></span><span class="op">(</span><span class="va">my_rf</span><span class="op">[</span>,<span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">27</span><span class="op">)</span><span class="op">]</span>, scale <span class="op">=</span> <span class="st">'column'</span>, trace <span class="op">=</span> <span class="st">'none'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="parameter-tuning-via-grid-search">Parameter Tuning via Grid Search<a class="anchor" aria-label="anchor" href="#parameter-tuning-via-grid-search"></a>
</h3>
<p>There are different optimization techniques to determine the
parameters for a random forest. A simple one to implement is a
grid-search where models with different parameters are initialized.
Based on their performance, the optimal parameters can be determined.
The parameters tested in our grid search are the number of principal
components included and the number of trees. Within the parameter values
of the number of principal components and number of trees,
cross-validation is done to determine the mtry variable.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">train_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>mtry <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>, ntrees <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>, num_pc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span>, acc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">train_proj_noids</span> <span class="op">&lt;-</span> <span class="va">train_proj_1000g</span><span class="op">[</span>,<span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">PC_inc</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw">for</span> <span class="op">(</span><span class="va">ntree</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span></span>
<span>    <span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">train</span><span class="op">(</span><span class="va">Pop</span><span class="op">~</span><span class="va">.</span>, data<span class="op">=</span><span class="va">train_proj_noids</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="va">PC_inc</span>,<span class="fl">21</span><span class="op">)</span><span class="op">]</span>, </span>
<span>                 method<span class="op">=</span><span class="st">"rf"</span>, metric<span class="op">=</span><span class="st">"Accuracy"</span>,</span>
<span>                 trControl<span class="op">=</span><span class="va">control</span>, ntree<span class="op">=</span><span class="va">ntree</span><span class="op">)</span></span>
<span>    <span class="va">train_proj_1000g</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">train_proj_1000g</span>, </span>
<span>                     <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>mtry <span class="op">=</span> <span class="va">fit</span><span class="op">$</span><span class="va">results</span><span class="op">$</span><span class="va">mtry</span>, </span>
<span>                                ntrees <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">ntree</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">results</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                                num_pc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="va">PC_inc</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">fit</span><span class="op">$</span><span class="va">results</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                                acc <span class="op">=</span> <span class="va">fit</span><span class="op">$</span><span class="va">results</span><span class="op">$</span><span class="va">Accuracy</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="evaluatinginterpretting-the-rf">Evaluating/Interpretting the RF<a class="anchor" aria-label="anchor" href="#evaluatinginterpretting-the-rf"></a>
</h3>
<p>Once you have found a successful OOB error rate, you can use the
below package to find out important information regarding your tree. The
below documentation builds a decision tree using categorical aspects to
classify the data. The PC values for each sample are provided in our
1000 Genomes reference dataset file, where each sample has values for 20
distinct PCs. We ask if a certain sample’s value for a specific PC is
less than or greater than a given number. Based on that answer, it is
either classified as a specific ancestry or needs to undergo more
investigation to classify it correctly. The visualization below is only
a subset of a decision tree the random forest builds as classifying all
26 ancestries builds quite a complicated and cluttered plot.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># This is a visual of the random forest being made. </span></span>
<span><span class="co"># https://github.com/araastat/reprtree</span></span>
<span><span class="fu">reprtree</span><span class="fu">:::</span><span class="fu">plot.getTree</span><span class="op">(</span><span class="va">rf_mtry</span><span class="op">)</span></span>
<span><span class="co"># Two different plots </span></span>
<span><span class="fu">reprtree</span><span class="fu">:::</span><span class="fu">plot.getTree</span><span class="op">(</span><span class="va">rf_mtry</span>, k<span class="op">=</span><span class="fl">1</span>, d<span class="op">=</span><span class="fl">5</span><span class="op">)</span> </span>
<span><span class="fu">reprtree</span><span class="fu">:::</span><span class="fu">plot.getTree</span><span class="op">(</span><span class="va">rf_mtry</span>, k<span class="op">=</span><span class="fl">1</span>, d<span class="op">=</span><span class="fl">6</span><span class="op">)</span> </span></code></pre></div>
</div>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body">
<div id="ref-Anderson2010" class="csl-entry">
<div class="csl-left-margin">1. </div>
<div class="csl-right-inline">Anderson CA, Pettersson FH, Clarke GM, Cardon
LR, Morris AP, Zondervan KT. <span class="nocase">Data quality control
in genetic case-control association studies.</span> Nature Protocols.
2010;5: 1564–73. doi:<a href="https://doi.org/10.1038/nprot.2010.116" class="external-link">10.1038/nprot.2010.116</a>
</div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Hannah Meyer.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
