<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Genotype quality control with plinkQC • plinkQC</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Genotype quality control with plinkQC">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="standard" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">plinkQC</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html" aria-label="ho"><span class="fa fa-home"></span></a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-documentation" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Documentation"><span class="fa fa-file"></span> Documentation</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-documentation">
<li><a class="dropdown-item" href="../articles/plinkQC.html">Step-by-step QC</a></li>
    <li><a class="dropdown-item" href="../articles/AncestryCheck.html">Ancestry</a></li>
    <li><a class="dropdown-item" href="../articles/HapMap.html">HapMap III reference</a></li>
    <li><a class="dropdown-item" href="../articles/Genomes1000.html">1000 Genomes reference</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="nav-link" href="../news/index.html"><span class="fa fa-newspaper-o"></span> news</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html"><span class="fa fa-code"></span> functions</a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/meyer-lab-cshl/plinkQC"><span class="fa fa-github fa-lg"></span> github</a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Genotype quality control with plinkQC</h1>
                        <h4 data-toc-skip class="author">Hannah Meyer
and Maha Syed</h4>
            
            <h4 data-toc-skip class="date">2025-11-20</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/meyer-lab-cshl/plinkQC/blob/HEAD/vignettes/plinkQC.Rmd" class="external-link"><code>vignettes/plinkQC.Rmd</code></a></small>
      <div class="d-none name"><code>plinkQC.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Genotyping arrays enable the direct measurement of an individuals
genotype at thousands of markers. Subsequent analyses such as
genome-wide association studies rely on the high quality of these marker
genotypes.</p>
<p>Anderson and colleagues introduced a protocol for data quality
control in genetic association studies heavily based on the summary
statistics and relatedness estimation functions in the PLINK software
<span class="citation">[1]</span>. PLINK is a comprehensive, open-source
command-line tool for genome-wide association studies (GWAS) and
population genetics research <span class="citation">[3]</span>. It’s
main functionalities include data management, computing individual- and
marker- level summary statistics, identity-by-state estimation and
association analysis.<br>
Integration with R is achieved through its R plugin or PLINK/SEQ R
Package <span class="citation">[4]</span>. While the plugin is limited
to operations yielding simple genetic marker vectors as output, the
PLINK/SEQ R Package is limited in the functionalities it can access.</p>
<p><em>plinkQC</em> facilitates genotype quality control for genetic
association studies as described by <span class="citation">[1]</span>.
It wraps around plink basic statistics (e.g. missing genotyping rates
per individual, allele frequencies per genetic marker) and relationship
functions and generates a per-individual and per-marker quality control
report. Individuals and markers that fail the quality control can
subsequently be removed with <em>plinkQC</em> to generate a new, clean
dataset. Removal of individuals based on relationship status is
optimised to retain as many individuals as possible in the study.</p>
<p>The majority if the functions in <em>plinkQC</em> depends on the <a href="https://www.cog-genomics.org/plink/1.9/" class="external-link">PLINK</a>
(<strong>version 1.9</strong>), which has to be manually installed prior
to the usage of <em>plinkQC</em>. The ancestry functions depends on the
newer version of <a href="https://www.cog-genomics.org/plink/2.0/" class="external-link">PLINK
2.0</a> (<strong>version 2.0</strong>). It assumes the genotype have
already been determined from the original probe intensity data of the
genotype array and is available in <a href="https://www.cog-genomics.org/plink/1.9/input#bed" class="external-link">plink
format</a>.</p>
<p>The protocol is implemented in four main functions, the
per-individual quality control (<code>perIndividualQC</code>), the
per-marker quality control (<code>perMarkerQC</code>). the generation of
the new, quality control dataset (<code>cleanData</code>), and the
prediction of human genetic ancestry
(<code>superpop_classification</code>):</p>
<div class="section level3">
<h3 id="per-individual-quality-control">Per-individual quality control<a class="anchor" aria-label="anchor" href="#per-individual-quality-control"></a>
</h3>
<p>The per-individual quality control with <code>perIndividualQC</code>
wraps around these functions: (i) <code>check_sex</code>: for the
identification of individuals with discordant sex information, (ii)
<code>check_heterozygosity_and_missingness</code>: for the
identification of individuals with outlying missing genotype and/or
heterozygosity rates, (iii) <code>check_relatedness</code>: for the
identification of related individuals (iv)
<code>ancestry_prediction</code>: for prediction of genomic ancestry</p>
</div>
<div class="section level3">
<h3 id="per-marker-quality-control">Per-marker quality control<a class="anchor" aria-label="anchor" href="#per-marker-quality-control"></a>
</h3>
<p>The per-marker quality control with <code>perMarkerQC</code> wraps
around these functions: (i) <code>check_snp_missingnes</code>: for the
identifying markers with excessive missing genotype rates, (ii)
<code>check_hwe</code>: for the identifying markers showing a
significant deviation from Hardy-Weinberg equilibrium (HWE), (iii)
<code>check_maf</code>: for the removal of markers with low minor allele
frequency (MAF).</p>
</div>
<div class="section level3">
<h3 id="clean-data">Clean data<a class="anchor" aria-label="anchor" href="#clean-data"></a>
</h3>
<p><code>cleanData</code> takes the results of <code>perMarkerQC</code>
and <code>perIndividualQC</code> and creates a new dataset with all
individuals and markers that passed the quality control checks.</p>
</div>
</div>
<div class="section level2">
<h2 id="workflow">Workflow<a class="anchor" aria-label="anchor" href="#workflow"></a>
</h2>
<p>In the following, genotype quality control with <em>plinkQC</em> is
applied on a small example dataset with 200 individuals and 10,000
markers (provided with this package). The quality control is
demonstrated in three easy steps, per-individual and per-marker quality
control followed by the generation of the new dataset. In addition, the
functionality of each of the functions underlying
<code>perMarkerQC</code> and <code>perIndividualQC</code> is
demonstrated at the end of this vignette.</p>
<p>To run the ancestry functionality of the package, <a href="https://www.cog-genomics.org/plink/2.0/" class="external-link">Plink v2</a> is needed.
Before use, the study data should be in the new hg38 annotation. USCS’s
<a href="https://genome.ucsc.edu/cgi-bin/hgLiftOver" class="external-link">liftOver</a> tool
may be needed to map variants from one annotation to another. More
details on how to use the tool can be found on the processing <a href="HapMap.html">HapMap III reference data vignette</a>. We provide an
example dataset that is in the hg38 annotation.</p>
<p>Additional loading matrices are needed for the PCA projection used in
the model. This is hosted on the plinkQC github repo under the
inst/extdata folder located <a href="https://github.com/meyer-lab-cshl/plinkQC/tree/ancestry_update/inst/extdata/load_mat" class="external-link">here</a>.
Alternatively, the whole github repo can downloaded with</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">git</span> clone git@github.com:meyer-lab-cshl/plinkQC.git</span></code></pre></div>
<p>The name of the files (before the .acount or .eigenvec.allele) for
the loading matrices must be included in the <code>path2load_mat</code>
variable.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">package.dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/find.package.html" class="external-link">find.package</a></span><span class="op">(</span><span class="st">'plinkQC'</span><span class="op">)</span></span>
<span><span class="va">indir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">package.dir</span>, <span class="st">'extdata'</span><span class="op">)</span></span>
<span><span class="va">qcdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">name</span> <span class="op">&lt;-</span> <span class="st">'data.hg38'</span></span>
<span><span class="va">path2plink</span> <span class="op">&lt;-</span> <span class="st">"/Users/syed/bin/plink"</span></span>
<span><span class="va">path2plink2</span> <span class="op">&lt;-</span> <span class="st">"/Users/syed/bin/plink2"</span></span>
<span><span class="va">path2load_mat</span> <span class="op">&lt;-</span> <span class="st">"path/to/load_mat/merged_chrs.postQC.train.pca"</span></span></code></pre></div>
<div class="section level3">
<h3 id="per-individual-quality-control-1">Per-individual quality control<a class="anchor" aria-label="anchor" href="#per-individual-quality-control-1"></a>
</h3>
<p>For <code>perIndividualQC</code>, one simply specifies the directory
where the data is stored (qcdir) and the prefix of the plink files
(i.e. prefix.bim, prefix.bed, prefix.fam). Per default, all quality
control checks will be conducted.</p>
<p>In addition to running each check, <code>perIndividualQC</code>
writes a list of all fail individual IDs to the qcdir. These IDs will be
removed in the computation of the <code>perMarkerQC</code>. If the list
is not present, <code>perMarkerQC</code> will send a message about
conducting the quality control on the entire dataset.</p>
<p>The ancestry portion of the package requires that the data should be
in PLINK 2.0 format (i.e pgen, pvar, psam files), and the variants
identifiers should be formatted similar to the following example:
1:12345[hg38]. <code>perIndividualQC</code> will run the functions
<code><a href="../reference/convert_to_plink2.html">convert_to_plink2()</a></code> and
<code><a href="../reference/rename_variant_identifiers.html">rename_variant_identifiers()</a></code> to convert the data into PLINK
2.0 format and rename the variant identifiers respectively.</p>
<p>NB: To reduce the data size of the example data in
<code>plinkQC</code>, data.genome has already been reduced to the
individuals that are related. Thus the relatedness plots in C only show
counts for related individuals only. For ancestry identification, there
are 597 SNPs that overlap between the study and reference data, roughly
0.2% of the reference dataset. Thus, additional noise is expected and
visible in the plot below.</p>
<p><code>perIndividualQC</code> displays the results of the quality
control steps in a multi-panel plot.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fail_individuals</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perIndividualQC.html">perIndividualQC</a></span><span class="op">(</span>indir<span class="op">=</span><span class="va">indir</span>, qcdir<span class="op">=</span><span class="va">qcdir</span>, name<span class="op">=</span><span class="va">name</span>,</span>
<span>                            path2plink<span class="op">=</span><span class="va">path2plink</span>, </span>
<span>                            interactive<span class="op">=</span><span class="cn">TRUE</span>, verbose<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="individualQC.png" alt="Individual quality control report" width="500px" style="display: block; margin: auto;"></p>
<p><code>overviewperIndividualQC</code> depicts overview plots of
quality control failures and the intersection of quality control
failures.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">overview_individuals</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/overviewPerIndividualQC.html">overviewPerIndividualQC</a></span><span class="op">(</span><span class="va">fail_individuals</span>,</span>
<span>                                                interactive<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="overviewQC.png" alt="Showing intersection of fails for individual quality  control with an upset plot" width="500px"></p>
</div>
<div class="section level3">
<h3 id="per-marker-quality-control-1">Per-marker quality control<a class="anchor" aria-label="anchor" href="#per-marker-quality-control-1"></a>
</h3>
<p><code>perMarkerQC</code> applies its checks to data in the specified
directory (qcdir), starting with the specified prefix of the plink files
(i.e. prefix.bim, prefix.bed, prefix.fam). Optionally, the user can
specify different thresholds for the quality control checks and which
check to conduct. Per default, all quality control checks will be
conducted. <code>perMarkerQC</code> displays the results of the QC step
in a multi-panel plot.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fail_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/perMarkerQC.html">perMarkerQC</a></span><span class="op">(</span>indir<span class="op">=</span><span class="va">indir</span>, qcdir<span class="op">=</span><span class="va">qcdir</span>, name<span class="op">=</span><span class="va">name</span>,</span>
<span>                            path2plink<span class="op">=</span><span class="va">path2plink</span>,</span>
<span>                            verbose<span class="op">=</span><span class="cn">TRUE</span>, interactive<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                            showPlinkOutput<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="markerQC.png" alt="Marker quality control reports" width="500px" style="display: block; margin: auto;"></p>
<p><code>overviewPerMarkerQC</code> depicts an overview of the marker
quality control failures and their overlaps.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">overview_marker</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/overviewPerMarkerQC.html">overviewPerMarkerQC</a></span><span class="op">(</span><span class="va">fail_markers</span>, interactive<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="overviewMarkerQC.png" alt="Overview of the marker quality control ran " width="500px"></p>
</div>
<div class="section level3">
<h3 id="create-qc-ed-dataset">Create QC-ed dataset<a class="anchor" aria-label="anchor" href="#create-qc-ed-dataset"></a>
</h3>
<p>After checking results of the per-individual and per-marker quality
control, individuals and markers that fail the chosen criteria can
automatically be removed from the dataset with <code>cleanData</code>,
resulting in the new dataset qcdir/data.clean.bed,qcdir/data.clean.bim,
qcdir/data.clean.fam. For convenience, <code>cleanData</code> returns a
list of all individuals in the study split into keep and remove
individuals.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">Ids</span>  <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cleanData.html">cleanData</a></span><span class="op">(</span>indir<span class="op">=</span><span class="va">indir</span>, qcdir<span class="op">=</span><span class="va">qcdir</span>, name<span class="op">=</span><span class="va">name</span>, path2plink<span class="op">=</span><span class="va">path2plink</span>,</span>
<span>                            verbose<span class="op">=</span><span class="cn">TRUE</span>, showPlinkOutput<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<div style="page-break-after: always;"></div>
</div>
</div>
<div class="section level2">
<h2 id="step-by-step">Step-by-step<a class="anchor" aria-label="anchor" href="#step-by-step"></a>
</h2>
<div class="section level3">
<h3 id="individuals-with-discordant-sex-information">Individuals with discordant sex information<a class="anchor" aria-label="anchor" href="#individuals-with-discordant-sex-information"></a>
</h3>
<p>The identification of individuals with discordant sex information
helps to detect sample mix-ups and samples with very poor genotyping
rates. For each sample, the homozygosity rates across all X-chromosomal
genetic markers are computed and compared with the expected rates
(typically $&lt;$0.2 for females and $&gt;$0.8 for males). For samples
where the assigned sex (PEDSEX in the .fam file) contradicts the sex
inferred from the homozygosity rates (SNPSEX), it should be checked that
the sex was correctly recorded (genotyping often occurs at different
locations as phenotyping and misrecording might occur). Samples with
discordant sex information that is not accounted for should be removed
from the study. Identifying individuals with discordant sex information
is implemented in <code>check_sex</code>. It finds individuals whose
SNPSEX != PEDSEX. Optionally, an extra data.frame with sample IDs and
sex can be provided to double check if external and PEDSEX data (often
processed at different centers) match. If a mismatch between PEDSEX and
SNPSEX was detected, by SNPSEX == Sex, PEDSEX of these individuals can
optionally be updated. <code>check_sex</code> depicts the X-chromosomal
heterozygosity (SNPSEX) of the samples split by their (PEDSEX).</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fail_sex</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/check_sex.html">check_sex</a></span><span class="op">(</span>indir<span class="op">=</span><span class="va">indir</span>, qcdir<span class="op">=</span><span class="va">qcdir</span>, name<span class="op">=</span><span class="va">name</span>, interactive<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                      verbose<span class="op">=</span><span class="cn">TRUE</span>, path2plink<span class="op">=</span><span class="va">path2plink</span><span class="op">)</span></span></code></pre></div>
<p><img src="checkSex.png" alt="Graph showing predicted sex versus labeled sex" width="500px" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="individuals-with-outlying-missing-genotype-andor-heterozygosity-rates">Individuals with outlying missing genotype and/or heterozygosity
rates<a class="anchor" aria-label="anchor" href="#individuals-with-outlying-missing-genotype-andor-heterozygosity-rates"></a>
</h3>
<p>The identification of individuals with outlying missing genotype
and/or heterozygosity rates helps to detect samples with poor DNA
quality and/or concentration that should be excluded from the study.
Typically, individuals with more than 3-7% of their genotype calls
missing are removed. Outlying heterozygosity rates are judged relative
to the overall heterozygosity rates in the study, and individuals whose
rates are more than a few standard deviations (sd) from the mean
heterozygosity rate are removed. A typical quality control for outlying
heterozygosity rates would remove individuals who are three sd away from
the mean rate. Identifying related individuals with outlying missing
genotype and/or heterozygosity rates is implemented in
<code>check_het_and_miss</code>. It finds individuals that have
genotyping and heterozygosity rates that fail the set thresholds and
depicts the results as a scatter plot with the samples’ missingness
rates on x-axis and their heterozygosity rates on the y-axis.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fail_het_imiss</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/check_het_and_miss.html">check_het_and_miss</a></span><span class="op">(</span>indir<span class="op">=</span><span class="va">indir</span>, qcdir<span class="op">=</span><span class="va">qcdir</span>, name<span class="op">=</span><span class="va">name</span>,</span>
<span>                                     interactive<span class="op">=</span><span class="cn">TRUE</span>, path2plink<span class="op">=</span><span class="va">path2plink</span><span class="op">)</span></span></code></pre></div>
<p><img src="checkHetImiss.png" alt="Showing the missingness on x-axis and heterzyogisity rate on the y-axis for individuals" width="500px" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="related-individuals">Related individuals<a class="anchor" aria-label="anchor" href="#related-individuals"></a>
</h3>
<p>Depending on the future use of the genotypes, it might required to
remove any related individuals from the study. Related individuals can
be identified by their proportion of shared alleles at the genotyped
markers (identity by descend, IBD). Standardly, individuals with
second-degree relatedness or higher will be excluded. Identifying
related individuals is implemented in <code>check_relatedness</code>. It
finds pairs of samples whose proportion of IBD is larger than the
specified highIBDTh. Subsequently, for pairs of individual that do not
have additional relatives in the dataset, the individual with the
greater genotype missingness rate is selected and returned as the
individual failing the relatedness check. For more complex family
structures, the unrelated individuals per family are selected (e.g. in a
parents-offspring trio, the offspring will be marked as fail, while the
parents will be kept in the analysis).</p>
<p>NB: To reduce the data size of the example data in
<code>plinkQC</code>, data.genome has already been reduced to the
individuals that are related. Thus the relatedness plots in C only show
counts for related individuals only.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">exclude_relatedness</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/check_relatedness.html">check_relatedness</a></span><span class="op">(</span>indir<span class="op">=</span><span class="va">indir</span>, qcdir<span class="op">=</span><span class="va">qcdir</span>, name<span class="op">=</span><span class="va">name</span>,</span>
<span>                                         interactive<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                                         path2plink<span class="op">=</span><span class="va">path2plink</span><span class="op">)</span></span></code></pre></div>
<p><img src="checkRelatedness.png" alt="bar graph with a histogram og IBD measurements" width="500px" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="ancestry-predictions-of-data">Ancestry Predictions of Data<a class="anchor" aria-label="anchor" href="#ancestry-predictions-of-data"></a>
</h3>
<p>For ancestry estimation, accessory functions from <a href="https://www.cog-genomics.org/plink/2.0/" class="external-link">Plink v2</a> are needed,
as are pre-computed loading matrices for the PCA projection used in the
model. These are hosted on the <em>plinkQC</em> github repo under the
inst/extdata folder located <a href="https://github.com/meyer-lab-cshl/plinkQC/tree/ancestry_update/inst/extdata/load_mat" class="external-link">here</a>.
Alternatively, the whole github repo can downloaded with</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="fu">git</span> clone git@github.com:meyer-lab-cshl/plinkQC.git</span></code></pre></div>
<p>For this, it is important that the data is in hg38 annotation. USCS’s
<a href="https://genome.ucsc.edu/cgi-bin/hgLiftOver" class="external-link">liftOver</a> tool
may be needed to map variants from one annotation to another. More
details on how to use the tool can be found on the processing <a href="HapMap.html">HapMap III reference data vignette</a>.</p>
<p>Additionally, the ancestry prediction function requires that the data
is in plink 2 format and has variant identifiers in the format:
1:12345[hg38]. The function <code>ancestry_prediction</code> will call
helper functions, <code><a href="../reference/convert_to_plink2.html">convert_to_plink2()</a></code> and
<code><a href="../reference/rename_variant_identifiers.html">rename_variant_identifiers()</a></code> to format the data as directed
by the parameters. The function returns: (i) prediction_prob: Dataframe
of family ids, sample ids, and model probability of the each ancestry.
(ii) prediction_majority: Dataframe of family ids, sample ids, and
ancestry with highest model probability. (iii) exclude_ancestry: A list
of ids to be excluded based on user-inputted ancestries to the ancestral
predictions of the samples (iv) p_ancestry: Bar graph of the ancestry
model probabilities.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">path2plink2</span> <span class="op">&lt;-</span> <span class="st">"/Users/syed/bin/plink2"</span></span>
<span><span class="va">path2load_mat</span> <span class="op">&lt;-</span> <span class="st">"path/to/load_mat/merged_chrs.postQC.train.pca"</span></span>
<span></span>
<span><span class="va">anc_check</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/ancestry_prediction.html">ancestry_prediction</a></span><span class="op">(</span>indir<span class="op">=</span><span class="va">indir</span>, qcdir<span class="op">=</span><span class="va">qcdir</span>, name<span class="op">=</span><span class="va">name</span>,</span>
<span>                                         interactive<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                                         path2plink2<span class="op">=</span><span class="va">path2plink2</span>, </span>
<span>                                         path2load_mat <span class="op">=</span> <span class="va">path2load_mat</span><span class="op">)</span></span></code></pre></div>
<p><img src="ancestral_prediction.png" alt="marker missingness" width="500px" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="markers-with-excessive-missingness-rate">Markers with excessive missingness rate<a class="anchor" aria-label="anchor" href="#markers-with-excessive-missingness-rate"></a>
</h3>
<p>Markers with excessive missingness rate are removed as they are
considered unreliable. Typically, thresholds for marker exclusion based
on missingness range from 1%-5%. Identifying markers with high
missingness rates is implemented in <code>snp_missingness</code>. It
calculates the rates of missing genotype calls and frequency for all
variants in the individuals that passed the
<code>perIndividualQC</code>.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fail_snpmissing</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/check_snp_missingness.html">check_snp_missingness</a></span><span class="op">(</span>indir<span class="op">=</span><span class="va">indir</span>, qcdir<span class="op">=</span><span class="va">qcdir</span>, name<span class="op">=</span><span class="va">name</span>,</span>
<span>                                         interactive<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                                         path2plink<span class="op">=</span><span class="va">path2plink</span>, </span>
<span>                                         showPlinkOutput<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="snpmissingness.png" alt="marker missingness" width="500px" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="markers-with-deviation-from-hwe">Markers with deviation from HWE<a class="anchor" aria-label="anchor" href="#markers-with-deviation-from-hwe"></a>
</h3>
<p>Markers with strong deviation from HWE might be indicative of
genotyping or genotype-calling errors. As serious genotyping errors
often yield very low p-values (in the order of
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msup><mn>10</mn><mo>−</mo></msup><mn>50</mn></mrow><annotation encoding="application/x-tex">10^-50</annotation></semantics></math>),
it is recommended to choose a reasonably low threshold to avoid
filtering too many variants (that might have slight, non-critical
deviations). Identifying markers with deviation from HWE is implemented
in <code>check_hwe</code>. It calculates the observed and expected
heterozygote frequencies per SNP in the individuals that passed the
<code>perIndividualQC</code> and computes the deviation of the
frequencies from Hardy-Weinberg equilibrium (HWE) by HWE exact test.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fail_hwe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/check_hwe.html">check_hwe</a></span><span class="op">(</span>indir<span class="op">=</span><span class="va">indir</span>, qcdir<span class="op">=</span><span class="va">qcdir</span>, name<span class="op">=</span><span class="va">name</span>, interactive<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                      path2plink<span class="op">=</span><span class="va">path2plink</span>, showPlinkOutput<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="hwe.png" alt="p-values of deviations from HWE histogram " width="500px" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="markers-with-low-minor-allele-frequency">Markers with low minor allele frequency<a class="anchor" aria-label="anchor" href="#markers-with-low-minor-allele-frequency"></a>
</h3>
<p>Markers with low minor allele count are often removed as the actual
genotype calling (via the calling algorithm) is very difficult due to
the small sizes of the heterozygote and rare-homozygote clusters.
Identifying markers with low minor allele count is implemented in
<code>check_maf</code>. It calculates the minor allele frequencies for
all variants in the individuals that passed the
<code>perIndividualQC</code>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fail_maf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/check_maf.html">check_maf</a></span><span class="op">(</span>indir<span class="op">=</span><span class="va">indir</span>, qcdir<span class="op">=</span><span class="va">qcdir</span>, name<span class="op">=</span><span class="va">name</span>, interactive<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>                      path2plink<span class="op">=</span><span class="va">path2plink</span>, showPlinkOutput<span class="op">=</span><span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p><img src="maf.png" alt="Histogram of minor allele frequencies" width="500px" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body">
<div id="ref-Anderson2010" class="csl-entry">
<div class="csl-left-margin">1. </div>
<div class="csl-right-inline">Anderson CA, Pettersson FH, Clarke GM, Cardon
LR, Morris AP, Zondervan KT. <span class="nocase">Data quality control
in genetic case-control association studies.</span> Nature Protocols.
2010;5: 1564–73. doi:<a href="https://doi.org/10.1038/nprot.2010.116" class="external-link">10.1038/nprot.2010.116</a>
</div>
</div>
<div id="ref-Purcell2007" class="csl-entry">
<div class="csl-left-margin">2. </div>
<div class="csl-right-inline">Purcell S, Neale B, Todd-Brown K, Thomas L,
Ferreira MAR, Bender D, et al. <span class="nocase">PLINK: a tool set
for whole-genome association and population-based linkage
analyses.</span> The American Journal of Human Genetics. 2007;81:
559–75. doi:<a href="https://doi.org/10.1086/519795" class="external-link">10.1086/519795</a>
</div>
</div>
<div id="ref-Chang2015" class="csl-entry">
<div class="csl-left-margin">3. </div>
<div class="csl-right-inline">Chang CC, Chow CC, Tellier LC, Vattikuti S,
Purcell SM, Lee JJ. <span class="nocase">Second-generation PLINK: rising
to the challenge of larger and richer datasets</span>. GigaScience.
2015;4: 7. doi:<a href="https://doi.org/10.1186/s13742-015-0047-8" class="external-link">10.1186/s13742-015-0047-8</a>
</div>
</div>
<div id="ref-PlinkSeq2014" class="csl-entry">
<div class="csl-left-margin">4. </div>
<div class="csl-right-inline">PLINK/SEQ. 2014. Available: <a href="https://atgu.mgh.harvard.edu/plinkseq/index.shtml" class="external-link">https://atgu.mgh.harvard.edu/plinkseq/index.shtml</a>
</div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Hannah Meyer.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
