<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Identification of SNPs with high missingness rate — check_snp_missingness • plinkQC</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Identification of SNPs with high missingness rate — check_snp_missingness"><meta property="og:description" content="Runs and evaluates results from plink --missing --freq. It calculate the
rates of missing genotype calls and frequency for all variants in the
individuals that passed the perIndividualQC. The SNP
missingness rates (stratified by minor allele frequency) are depicted as
histograms."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">plinkQC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/plinkQC.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/AncestryCheck.html">Ancestry estimation based on reference samples of known ethnicities</a>
    </li>
    <li>
      <a href="../articles/Genomes1000.html">Processing 1000 Genomes reference data for ancestry estimation</a>
    </li>
    <li>
      <a href="../articles/HapMap.html">Processing HapMap III reference data for ancestry estimation</a>
    </li>
  </ul></li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/meyer-lab-cshl/plinkQC/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Identification of SNPs with high missingness rate</h1>
    <small class="dont-index">Source: <a href="https://github.com/meyer-lab-cshl/plinkQC/blob/HEAD/R/markerQC.R" class="external-link"><code>R/markerQC.R</code></a></small>
    <div class="hidden name"><code>check_snp_missingness.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Runs and evaluates results from plink --missing --freq. It calculate the
rates of missing genotype calls and frequency for all variants in the
individuals that passed the <code><a href="perIndividualQC.html">perIndividualQC</a></code>. The SNP
missingness rates (stratified by minor allele frequency) are depicted as
histograms.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">check_snp_missingness</span><span class="op">(</span></span>
<span>  <span class="va">indir</span>,</span>
<span>  <span class="va">name</span>,</span>
<span>  qcdir <span class="op">=</span> <span class="va">indir</span>,</span>
<span>  lmissTh <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  interactive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  path2plink <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  showPlinkOutput <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  keep_individuals <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  remove_individuals <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  exclude_markers <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  extract_markers <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  legend_text_size <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  legend_title_size <span class="op">=</span> <span class="fl">7</span>,</span>
<span>  axis_text_size <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  axis_title_size <span class="op">=</span> <span class="fl">7</span>,</span>
<span>  title_size <span class="op">=</span> <span class="fl">9</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>indir</dt>
<dd><p>[character] /path/to/directory containing the basic PLINK data
files name.bim, name.bed, name.fam files.</p></dd>


<dt>name</dt>
<dd><p>[character] Prefix of PLINK files, i.e. name.bed, name.bim,
name.fam.</p></dd>


<dt>qcdir</dt>
<dd><p>[character] /path/to/directory where results will be written to.
If <code><a href="perIndividualQC.html">perIndividualQC</a></code> was conducted, this directory should be the
same as qcdir specified in <code><a href="perIndividualQC.html">perIndividualQC</a></code>, i.e. it contains
name.fail.IDs with IIDs of individuals that failed QC. User needs writing
permission to qcdir. Per default, qcdir=indir.</p></dd>


<dt>lmissTh</dt>
<dd><p>[double] Threshold for acceptable variant missing rate across
samples.</p></dd>


<dt>interactive</dt>
<dd><p>[logical] Should plots be shown interactively? When
choosing this option, make sure you have X-forwarding/graphical interface
available for interactive plotting. Alternatively, set interactive=FALSE and
save the returned plot object (p_lmiss) via ggplot2::ggsave(p=p_lmiss,
other_arguments) or pdf(outfile) print(p_lmiss) dev.off().</p></dd>


<dt>path2plink</dt>
<dd><p>[character] Absolute path to PLINK executable
(<a href="https://www.cog-genomics.org/plink/1.9/" class="external-link">https://www.cog-genomics.org/plink/1.9/</a>) i.e.
plink should be accessible as path2plink -h. The full name of the executable
should be specified: for windows OS, this means path/plink.exe, for unix
platforms this is path/plink. If not provided, assumed that PATH set-up works
and PLINK will be found by <code><a href="https://rdrr.io/pkg/sys/man/exec.html" class="external-link">exec</a></code>('plink').</p></dd>


<dt>verbose</dt>
<dd><p>[logical] If TRUE, progress info is printed to standard out
and specifically, if TRUE, plink log will be displayed.</p></dd>


<dt>showPlinkOutput</dt>
<dd><p>[logical] If TRUE, plink log and error messages are
printed to standard out.</p></dd>


<dt>keep_individuals</dt>
<dd><p>[character] Path to file with individuals to be
retained in the analysis. The file has to be a space/tab-delimited text file
with family IDs in the first column and within-family IDs in the second
column. All samples not listed in this file will be removed from the current
analysis. See <a href="https://www.cog-genomics.org/plink/1.9/filter#indiv" class="external-link">https://www.cog-genomics.org/plink/1.9/filter#indiv</a>.
Default: NULL, i.e. no filtering on individuals.</p></dd>


<dt>remove_individuals</dt>
<dd><p>[character] Path to file with individuals to be
removed from the analysis. The file has to be a space/tab-delimited text file
with family IDs in the first column and within-family IDs in the second
column. All samples listed in this file will be removed from the current
analysis. See <a href="https://www.cog-genomics.org/plink/1.9/filter#indiv" class="external-link">https://www.cog-genomics.org/plink/1.9/filter#indiv</a>.
Default: NULL, i.e. no filtering on individuals.</p></dd>


<dt>exclude_markers</dt>
<dd><p>[character] Path to file with makers to be
removed from the analysis. The file has to be a text file with a list of
variant IDs (usually one per line, but it's okay for them to just be
separated by spaces). All listed variants will be removed from the current
analysis. See <a href="https://www.cog-genomics.org/plink/1.9/filter#snp" class="external-link">https://www.cog-genomics.org/plink/1.9/filter#snp</a>.
Default: NULL, i.e. no filtering on markers.</p></dd>


<dt>extract_markers</dt>
<dd><p>[character] Path to file with makers to be
included in the analysis. The file has to be a text file with a list of
variant IDs (usually one per line, but it's okay for them to just be
separated by spaces). All unlisted variants will be removed from the current
analysis. See <a href="https://www.cog-genomics.org/plink/1.9/filter#snp" class="external-link">https://www.cog-genomics.org/plink/1.9/filter#snp</a>.
Default: NULL, i.e. no filtering on markers.</p></dd>


<dt>legend_text_size</dt>
<dd><p>[integer] Size for legend text.</p></dd>


<dt>legend_title_size</dt>
<dd><p>[integer] Size for legend title.</p></dd>


<dt>axis_text_size</dt>
<dd><p>[integer] Size for axis text.</p></dd>


<dt>axis_title_size</dt>
<dd><p>[integer] Size for axis title.</p></dd>


<dt>title_size</dt>
<dd><p>[integer] Size for plot title.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>Named list with i) fail_missingness containing a [data.frame] with
CHR (Chromosome code), SNP (Variant identifier), CLST (Cluster identifier.
Only present with --within/--family), N_MISS (Number of missing genotype
call(s), not counting obligatory missings), N_CLST  (Cluster size; does not
include nonmales on Ychr; Only present with --within/--family), N_GENO
(Number of potentially valid call(s)), F_MISS (Missing call rate) for all
SNPs failing the lmissTh and ii) p_lmiss, a ggplot2-object 'containing' the
SNP missingness histogram which can be shown by (print(p_lmiss)).</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p><code>check_snp_missingness</code> uses plink --remove name.fail.IDs --missing
--freq to calculate rates of missing genotype calls and frequency per SNP in
the individuals that passed the <code><a href="perIndividualQC.html">perIndividualQC</a></code>. It does so
without generating a new dataset but simply removes the IDs when calculating
the statistics.</p>
<p>For details on the output data.frame fail_missingness, check the original
description on the PLINK output format page:
<a href="https://www.cog-genomics.org/plink/1.9/formats#lmiss" class="external-link">https://www.cog-genomics.org/plink/1.9/formats#lmiss</a>.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">indir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package<span class="op">=</span><span class="st">"plinkQC"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">qcdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">name</span> <span class="op">&lt;-</span> <span class="st">"data"</span></span></span>
<span class="r-in"><span><span class="va">path2plink</span> <span class="op">&lt;-</span> <span class="st">'/path/to/plink'</span></span></span>
<span class="r-in"><span><span class="co"># the following code is not run on package build, as the path2plink on the</span></span></span>
<span class="r-in"><span><span class="co"># user system is not known.</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co"># run on all individuals and markers</span></span></span>
<span class="r-in"><span><span class="va">fail_snp_missingness</span> <span class="op">&lt;-</span> <span class="fu">check_snp_missingness</span><span class="op">(</span>qcdir<span class="op">=</span><span class="va">qcdir</span>, indir<span class="op">=</span><span class="va">indir</span>,</span></span>
<span class="r-in"><span>name<span class="op">=</span><span class="va">name</span>, interactive<span class="op">=</span><span class="cn">FALSE</span>, verbose<span class="op">=</span><span class="cn">TRUE</span>, path2plink<span class="op">=</span><span class="va">path2plink</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># run on subset of individuals and markers</span></span></span>
<span class="r-in"><span><span class="va">keep_individuals_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"keep_individuals"</span>,</span></span>
<span class="r-in"><span>package<span class="op">=</span><span class="st">"plinkQC"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">extract_markers_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"extract_markers"</span>,</span></span>
<span class="r-in"><span>package<span class="op">=</span><span class="st">"plinkQC"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fail_snp_missingness</span> <span class="op">&lt;-</span> <span class="fu">check_snp_missingness</span><span class="op">(</span>qcdir<span class="op">=</span><span class="va">qcdir</span>, indir<span class="op">=</span><span class="va">indir</span>,</span></span>
<span class="r-in"><span>name<span class="op">=</span><span class="va">name</span>, interactive<span class="op">=</span><span class="cn">FALSE</span>, verbose<span class="op">=</span><span class="cn">TRUE</span>, path2plink<span class="op">=</span><span class="va">path2plink</span>,</span></span>
<span class="r-in"><span>keep_individuals<span class="op">=</span><span class="va">keep_individuals_file</span>, extract_markers<span class="op">=</span><span class="va">extract_markers_file</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Hannah Meyer, Caroline Walter.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

