<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Create plink dataset with individuals and markers passing quality control — cleanData • plinkQC</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Create plink dataset with individuals and markers passing quality control — cleanData"><meta property="og:description" content="Individuals that fail per-individual QC and markers that fail
per-marker QC are removed from indir/name.bim/.bed/.fam and a new, dataset
with the remaining individuals and markers is created as
qcdir/name.clean.bim/.bed/.fam."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-standard navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">plinkQC</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.3.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../index.html">
    <span class="fa fa-home"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    <span class="fa fa-file"></span>
     
    documentation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/plinkQC.html">Step-by-step QC</a>
    </li>
    <li>
      <a href="../articles/AncestryCheck.html">Ancestry</a>
    </li>
    <li>
      <a href="../articles/HapMap.html">HapMap III reference</a>
    </li>
    <li>
      <a href="../articles/Genomes1000.html">1000 Genomes reference</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="../news/index.html">
    <span class="fa fa-newspaper-o"></span>
     
    news
  </a>
</li>
<li>
  <a href="../reference/index.html">
    <span class="fa fa-code"></span>
     
    functions
  </a>
</li>
<li>
  <a href="https://github.com/meyer-lab-cshl/plinkQC" class="external-link">
    <span class="fa fa-github fa-lg"></span>
     
    github
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Create plink dataset with individuals and markers passing quality control</h1>
    <small class="dont-index">Source: <a href="https://github.com/meyer-lab-cshl/plinkQC/blob/HEAD/R/applyQC.R" class="external-link"><code>R/applyQC.R</code></a></small>
    <div class="hidden name"><code>cleanData.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Individuals that fail per-individual QC and markers that fail
per-marker QC are removed from indir/name.bim/.bed/.fam and a new, dataset
with the remaining individuals and markers is created as
qcdir/name.clean.bim/.bed/.fam.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">cleanData</span><span class="op">(</span></span>
<span>  <span class="va">indir</span>,</span>
<span>  <span class="va">name</span>,</span>
<span>  qcdir <span class="op">=</span> <span class="va">indir</span>,</span>
<span>  filterSex <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  filterHeterozygosity <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  filterSampleMissingness <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  filterAncestry <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  filterRelated <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  filterSNPMissingness <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  lmissTh <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  filterHWE <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  hweTh <span class="op">=</span> <span class="fl">1e-05</span>,</span>
<span>  filterMAF <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  macTh <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  mafTh <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  path2plink <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  keep_individuals <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  remove_individuals <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  exclude_markers <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  extract_markers <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  showPlinkOutput <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>indir</dt>
<dd><p>[character] /path/to/directory containing the basic PLINK data
files name.bim, name.bed, name.fam files.</p></dd>


<dt>name</dt>
<dd><p>[character] Prefix of PLINK files, i.e. name.bed, name.bim,
name.fam.</p></dd>


<dt>qcdir</dt>
<dd><p>[character] /path/to/directory where results will be written to.
If <code><a href="perIndividualQC.html">perIndividualQC</a></code> was conducted, this directory should be the
same as qcdir specified in <code><a href="perIndividualQC.html">perIndividualQC</a></code>, i.e. it contains
name.fail.IDs with IIDs of individuals that failed QC. User needs writing
permission to qcdir. Per default, qcdir=indir.</p></dd>


<dt>filterSex</dt>
<dd><p>[logical] Set to exclude samples that failed the sex
check (via <code><a href="check_sex.html">check_sex</a></code> or <code><a href="perIndividualQC.html">perIndividualQC</a></code>).
Requires file qcdir/name.fail-sexcheck.IDs (automatically created by
<code><a href="perIndividualQC.html">perIndividualQC</a></code> if do.evaluate_check_sex set to TRUE).</p></dd>


<dt>filterHeterozygosity</dt>
<dd><p>[logical] Set to exclude samples that failed
check for outlying heterozygosity rates (via
<code><a href="check_het_and_miss.html">check_het_and_miss</a></code> or
<code><a href="perIndividualQC.html">perIndividualQC</a></code>). Requires file qcdir/name.fail-het.IDs
(automatically created by <code><a href="perIndividualQC.html">perIndividualQC</a></code> if
do.evaluate_check_het_and_miss set to TRUE).</p></dd>


<dt>filterSampleMissingness</dt>
<dd><p>[logical] Set to exclude samples that failed
check for excessive missing genotype rates (via
<code><a href="check_het_and_miss.html">check_het_and_miss</a></code> or
<code><a href="perIndividualQC.html">perIndividualQC</a></code>). Requires file qcdir/name.fail-imiss.IDs
(automatically created by <code><a href="perIndividualQC.html">perIndividualQC</a></code> if
do.evaluate_check_het_and_miss set to TRUE).</p></dd>


<dt>filterAncestry</dt>
<dd><p>[logical] Set to exclude samples that failed ancestry
check (via <code><a href="check_ancestry.html">check_ancestry</a></code> or <code><a href="perIndividualQC.html">perIndividualQC</a></code>).
Requires file qcdir/name.fail-ancestry.IDs (automatically created by
<code><a href="perIndividualQC.html">perIndividualQC</a></code> if do.check_ancestry set to TRUE).</p></dd>


<dt>filterRelated</dt>
<dd><p>[logical] Set to exclude samples that failed relatedness
check (via <code><a href="check_relatedness.html">check_relatedness</a></code> or <code><a href="perIndividualQC.html">perIndividualQC</a></code>).
Requires file qcdir/name.fail-IBD.IDs (automatically created by
<code><a href="perIndividualQC.html">perIndividualQC</a></code> if do.evaluate_check_relatedness set to TRUE).</p></dd>


<dt>filterSNPMissingness</dt>
<dd><p>[logical] Set to exclude markers that have
excessive missing rates across samples (via
<code><a href="check_snp_missingness.html">check_snp_missingness</a></code> or <code><a href="perMarkerQC.html">perMarkerQC</a></code>). Requires
lmissTh to be set.</p></dd>


<dt>lmissTh</dt>
<dd><p>[double] Threshold for acceptable variant missing rate across
samples.</p></dd>


<dt>filterHWE</dt>
<dd><p>[logical] Set to exclude markers that fail HWE exact test
(via <code><a href="check_hwe.html">check_hwe</a></code> or <code><a href="perMarkerQC.html">perMarkerQC</a></code>). Requires hweTh to
be set.</p></dd>


<dt>hweTh</dt>
<dd><p>[double] Significance threshold for deviation from HWE.</p></dd>


<dt>filterMAF</dt>
<dd><p>[logical] Set to exclude markers that fail minor allele
frequency or minor allele count threshold (via <code><a href="check_maf.html">check_maf</a></code> or
<code><a href="perMarkerQC.html">perMarkerQC</a></code>). Requires mafTh or macTh to be set.</p></dd>


<dt>macTh</dt>
<dd><p>[double] Threshold for minor allele cut cut-off, if both mafTh
and macTh are specified, macTh is used (macTh = mafTh\*2\*NrSamples).</p></dd>


<dt>mafTh</dt>
<dd><p>[double] Threshold for minor allele frequency cut-off.</p></dd>


<dt>path2plink</dt>
<dd><p>[character] Absolute path to PLINK executable
(<a href="https://www.cog-genomics.org/plink/1.9/" class="external-link">https://www.cog-genomics.org/plink/1.9/</a>) i.e.
plink should be accessible as path2plink -h. The full name of the executable
should be specified: for windows OS, this means path/plink.exe, for unix
platforms this is path/plink. If not provided, assumed that PATH set-up works
and PLINK will be found by <code><a href="https://rdrr.io/pkg/sys/man/exec.html" class="external-link">exec</a></code>('plink').</p></dd>


<dt>verbose</dt>
<dd><p>[logical] If TRUE, progress info is printed to standard out.</p></dd>


<dt>keep_individuals</dt>
<dd><p>[character] Path to file with individuals to be
retained in the analysis. The file has to be a space/tab-delimited text file
with family IDs in the first column and within-family IDs in the second
column. All samples not listed in this file will be removed from the current
analysis. See <a href="https://www.cog-genomics.org/plink/1.9/filter#indiv" class="external-link">https://www.cog-genomics.org/plink/1.9/filter#indiv</a>.
Default: NULL, i.e. no filtering on individuals.</p></dd>


<dt>remove_individuals</dt>
<dd><p>[character] Path to file with individuals to be
removed from the analysis. The file has to be a space/tab-delimited text file
with family IDs in the first column and within-family IDs in the second
column. All samples listed in this file will be removed from the current
analysis. See <a href="https://www.cog-genomics.org/plink/1.9/filter#indiv" class="external-link">https://www.cog-genomics.org/plink/1.9/filter#indiv</a>.
Default: NULL, i.e. no filtering on individuals.</p></dd>


<dt>exclude_markers</dt>
<dd><p>[character] Path to file with makers to be
removed from the analysis. The file has to be a text file with a list of
variant IDs (usually one per line, but it's okay for them to just be
separated by spaces). All listed variants will be removed from the current
analysis. See <a href="https://www.cog-genomics.org/plink/1.9/filter#snp" class="external-link">https://www.cog-genomics.org/plink/1.9/filter#snp</a>.
Default: NULL, i.e. no filtering on markers.</p></dd>


<dt>extract_markers</dt>
<dd><p>[character] Path to file with makers to be
included in the analysis. The file has to be a text file with a list of
variant IDs (usually one per line, but it's okay for them to just be
separated by spaces). All unlisted variants will be removed from the current
analysis. See <a href="https://www.cog-genomics.org/plink/1.9/filter#snp" class="external-link">https://www.cog-genomics.org/plink/1.9/filter#snp</a>.
Default: NULL, i.e. no filtering on markers.</p></dd>


<dt>showPlinkOutput</dt>
<dd><p>[logical] If TRUE, plink log and error messages are
printed to standard out.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>names [list] with i) passIDs, containing a [data.frame] with family
[FID] and individual [IID] IDs of samples that pass the QC, ii) failIDs,
containing a [data.frame] with family [FID] and individual [IID] IDs of
samples that fail the QC.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">package.dir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/find.package.html" class="external-link">find.package</a></span><span class="op">(</span><span class="st">'plinkQC'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">indir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">package.dir</span>, <span class="st">'extdata'</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">qcdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">name</span> <span class="op">&lt;-</span> <span class="st">"data"</span></span></span>
<span class="r-in"><span><span class="va">path2plink</span> <span class="op">&lt;-</span> <span class="st">'/path/to/plink'</span></span></span>
<span class="r-in"><span><span class="co"># the following code is not run on package build, as the path2plink on the</span></span></span>
<span class="r-in"><span><span class="co"># user system is not known.</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co"># Run qc on all samples and markers in the dataset</span></span></span>
<span class="r-in"><span><span class="co">## Run individual QC checks</span></span></span>
<span class="r-in"><span><span class="va">fail_individuals</span> <span class="op">&lt;-</span> <span class="fu"><a href="perIndividualQC.html">perIndividualQC</a></span><span class="op">(</span>indir<span class="op">=</span><span class="va">indir</span>, qcdir<span class="op">=</span><span class="va">qcdir</span>, name<span class="op">=</span><span class="va">name</span>,</span></span>
<span class="r-in"><span>refSamplesFile<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">qcdir</span>, <span class="st">"/HapMap_ID2Pop.txt"</span>,sep<span class="op">=</span><span class="st">""</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>refColorsFile<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">qcdir</span>, <span class="st">"/HapMap_PopColors.txt"</span>, sep<span class="op">=</span><span class="st">""</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>prefixMergedDataset<span class="op">=</span><span class="st">"data.HapMapIII"</span>, interactive<span class="op">=</span><span class="cn">FALSE</span>, verbose<span class="op">=</span><span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>path2plink<span class="op">=</span><span class="va">path2plink</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Run marker QC checks</span></span></span>
<span class="r-in"><span><span class="va">fail_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="perMarkerQC.html">perMarkerQC</a></span><span class="op">(</span>indir<span class="op">=</span><span class="va">indir</span>, qcdir<span class="op">=</span><span class="va">qcdir</span>, name<span class="op">=</span><span class="va">name</span>,</span></span>
<span class="r-in"><span>path2plink<span class="op">=</span><span class="va">path2plink</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Create new dataset of individuals and markers passing QC</span></span></span>
<span class="r-in"><span><span class="va">ids_all</span> <span class="op">&lt;-</span> <span class="fu">cleanData</span><span class="op">(</span>indir<span class="op">=</span><span class="va">indir</span>, qcdir<span class="op">=</span><span class="va">qcdir</span>, name<span class="op">=</span><span class="va">name</span>, macTh<span class="op">=</span><span class="fl">15</span>,</span></span>
<span class="r-in"><span>verbose<span class="op">=</span><span class="cn">TRUE</span>, path2plink<span class="op">=</span><span class="va">path2plink</span>, filterAncestry<span class="op">=</span><span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>filterRelated<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Run qc on subset of samples and markers in the dataset</span></span></span>
<span class="r-in"><span><span class="va">highlight_samples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"keep_individuals"</span>,</span></span>
<span class="r-in"><span>package<span class="op">=</span><span class="st">"plinkQC"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">remove_individuals_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"remove_individuals"</span>,</span></span>
<span class="r-in"><span>package<span class="op">=</span><span class="st">"plinkQC"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">fail_individuals</span> <span class="op">&lt;-</span> <span class="fu"><a href="perIndividualQC.html">perIndividualQC</a></span><span class="op">(</span>indir<span class="op">=</span><span class="va">indir</span>, qcdir<span class="op">=</span><span class="va">qcdir</span>, name<span class="op">=</span><span class="va">name</span>,</span></span>
<span class="r-in"><span>dont.check_ancestry <span class="op">=</span> <span class="cn">TRUE</span>, interactive<span class="op">=</span><span class="cn">FALSE</span>, verbose<span class="op">=</span><span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>highlight_samples <span class="op">=</span> <span class="va">highlight_samples</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>, highlight_type <span class="op">=</span> <span class="st">"label"</span>,</span></span>
<span class="r-in"><span>remove_individuals <span class="op">=</span> <span class="va">remove_individuals_file</span>, path2plink<span class="op">=</span><span class="va">path2plink</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Run marker QC checks</span></span></span>
<span class="r-in"><span><span class="va">fail_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="perMarkerQC.html">perMarkerQC</a></span><span class="op">(</span>indir<span class="op">=</span><span class="va">indir</span>, qcdir<span class="op">=</span><span class="va">qcdir</span>, name<span class="op">=</span><span class="va">name</span>,</span></span>
<span class="r-in"><span>path2plink<span class="op">=</span><span class="va">path2plink</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Create new dataset of individuals and markers passing QC</span></span></span>
<span class="r-in"><span><span class="va">ids_all</span> <span class="op">&lt;-</span> <span class="fu">cleanData</span><span class="op">(</span>indir<span class="op">=</span><span class="va">indir</span>, qcdir<span class="op">=</span><span class="va">qcdir</span>, name<span class="op">=</span><span class="va">name</span>, macTh<span class="op">=</span><span class="fl">15</span>,</span></span>
<span class="r-in"><span>verbose<span class="op">=</span><span class="cn">TRUE</span>, path2plink<span class="op">=</span><span class="va">path2plink</span>, filterAncestry<span class="op">=</span><span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>remove_individuals <span class="op">=</span> <span class="va">remove_individuals_file</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Hannah Meyer.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

