<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Package index â€¢ plinkQC</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Package index"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="standard" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">plinkQC</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html" aria-label="ho"><span class="fa fa-home"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-documentation" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Documentation"><span class="fa fa-file"></span> Documentation</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-documentation"><li><a class="dropdown-item" href="../articles/plinkQC.html">Step-by-step QC</a></li>
    <li><a class="dropdown-item" href="../articles/AncestryCheck.html">Ancestry</a></li>
    <li><a class="dropdown-item" href="../articles/HapMap.html">HapMap III reference</a></li>
    <li><a class="dropdown-item" href="../articles/Genomes1000.html">1000 Genomes reference</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="nav-link" href="../news/index.html"><span class="fa fa-newspaper-o"></span> news</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html"><span class="fa fa-code"></span> functions</a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/meyer-lab-cshl/plinkQC"><span class="fa fa-github fa-lg"></span> github</a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Package index</h1>
    </div>

    <div class="section level2">
      <h2 id="main-quality-control-functions">Main quality control functions<a class="anchor" aria-label="anchor" href="#main-quality-control-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="perIndividualQC.html">perIndividualQC()</a></code>

        </dt>
        <dd>Quality control for all individuals in plink-dataset</dd>
      </dl><dl><dt>

          <code><a href="perMarkerQC.html">perMarkerQC()</a></code>

        </dt>
        <dd>Quality control for all markers in plink-dataset</dd>
      </dl><dl><dt>

          <code><a href="overviewPerIndividualQC.html">overviewPerIndividualQC()</a></code>

        </dt>
        <dd>Overview of per sample QC</dd>
      </dl><dl><dt>

          <code><a href="overviewPerMarkerQC.html">overviewPerMarkerQC()</a></code>

        </dt>
        <dd>Overview of per marker QC</dd>
      </dl><dl><dt>

          <code><a href="cleanData.html">cleanData()</a></code>

        </dt>
        <dd>Create plink dataset with individuals and markers passing quality control</dd>
      </dl></div><div class="section level2">
      <h2 id="individual-quality-control-functions">Individual quality control functions<a class="anchor" aria-label="anchor" href="#individual-quality-control-functions"></a></h2>

      <div class="section-desc"><p>Functions for step-by-step per-individual quality control</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="check_sex.html">check_sex()</a></code>

        </dt>
        <dd>Identification of individuals with discordant sex information</dd>
      </dl><dl><dt>

          <code><a href="check_relatedness.html">check_relatedness()</a></code>

        </dt>
        <dd>Identification of related individuals</dd>
      </dl><dl><dt>

          <code><a href="check_het_and_miss.html">check_het_and_miss()</a></code>

        </dt>
        <dd>Identification of individuals with outlying missing genotype or heterozygosity rates</dd>
      </dl></div><div class="section level2">
      <h2 id="marker-quality-control-functions">Marker quality control functions<a class="anchor" aria-label="anchor" href="#marker-quality-control-functions"></a></h2>

      <div class="section-desc"><p>Functions for step-by-step per-marker quality control</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="check_snp_missingness.html">check_snp_missingness()</a></code>

        </dt>
        <dd>Identification of SNPs with high missingness rate</dd>
      </dl><dl><dt>

          <code><a href="check_maf.html">check_maf()</a></code>

        </dt>
        <dd>Identification of SNPs with low minor allele frequency</dd>
      </dl><dl><dt>

          <code><a href="check_hwe.html">check_hwe()</a></code>

        </dt>
        <dd>Identification of SNPs showing a significant deviation from Hardy-Weinberg- equilibrium (HWE)</dd>
      </dl></div><div class="section level2">
      <h2 id="quality-control-helper-functions">Quality control helper functions<a class="anchor" aria-label="anchor" href="#quality-control-helper-functions"></a></h2>

      <div class="section-desc"><p>Helper functions for step-by-step per-individual quality control: accesible to the user, but recommended use via per-individual check_* functions.</p></div>


    </div><div class="section level2">




      <dl><dt>

          <code><a href="run_check_sex.html">run_check_sex()</a></code>

        </dt>
        <dd>Run PLINK sexcheck</dd>
      </dl><dl><dt>

          <code><a href="evaluate_check_sex.html">evaluate_check_sex()</a></code>

        </dt>
        <dd>Evaluate results from PLINK sex check.</dd>
      </dl><dl><dt>

          <code><a href="run_check_heterozygosity.html">run_check_heterozygosity()</a></code>

        </dt>
        <dd>Run PLINK heterozygosity rate calculation</dd>
      </dl><dl><dt>

          <code><a href="run_check_missingness.html">run_check_missingness()</a></code>

        </dt>
        <dd>Run PLINK missingness rate calculation</dd>
      </dl><dl><dt>

          <code><a href="evaluate_check_het_and_miss.html">evaluate_check_het_and_miss()</a></code>

        </dt>
        <dd>Evaluate results from PLINK missing genotype and heterozygosity rate check.</dd>
      </dl><dl><dt>

          <code><a href="run_check_relatedness.html">run_check_relatedness()</a></code>

        </dt>
        <dd>Run PLINK IBD estimation</dd>
      </dl><dl><dt>

          <code><a href="evaluate_check_relatedness.html">evaluate_check_relatedness()</a></code>

        </dt>
        <dd>Evaluate results from PLINK IBD estimation.</dd>
      </dl></div><div class="section level2">
      <h2 id="general-helper-functions">General helper functions<a class="anchor" aria-label="anchor" href="#general-helper-functions"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="checkPlink.html">checkPlink()</a></code>

        </dt>
        <dd>Check PLINK software access</dd>
      </dl><dl><dt>

          <code><a href="checkPlink2.html">checkPlink2()</a></code>

        </dt>
        <dd>Check PLINK2 software access</dd>
      </dl><dl><dt>

          <code><a href="testNumerics.html">testNumerics()</a></code>

        </dt>
        <dd>Test lists for different properties of numerics</dd>
      </dl><dl><dt>

          <code><a href="relatednessFilter.html">relatednessFilter()</a></code>

        </dt>
        <dd>Remove related individuals while keeping maximum number of individuals</dd>
      </dl><dl><dt>

          <code><a href="checkFiltering.html">checkFiltering()</a></code>

        </dt>
        <dd>Check and construct PLINK sample and marker filters</dd>
      </dl><dl><dt>

          <code><a href="rename_variant_identifiers.html">rename_variant_identifiers()</a></code>

        </dt>
        <dd>Renaming variants</dd>
      </dl><dl><dt>

          <code><a href="pruning_ld.html">pruning_ld()</a></code>

        </dt>
        <dd>Pruning of SNPs in Linkage Disequilibrium</dd>
      </dl></div><div class="section level2">
      <h2 id="identifying-ancestry">Identifying Ancestry<a class="anchor" aria-label="anchor" href="#identifying-ancestry"></a></h2>




    </div><div class="section level2">




      <dl><dt>

          <code><a href="convert_to_plink2.html">convert_to_plink2()</a></code>

        </dt>
        <dd>Converting PLINK v1.9 data files into PLINK v2.0 data files</dd>
      </dl><dl><dt>

          <code><a href="superpop_classification.html">superpop_classification()</a></code>

        </dt>
        <dd>Predicting sample superpopulation ancestry</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Hannah Meyer.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

