<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Quality control for all markers in plink-dataset — perMarkerQC • plinkQC</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Quality control for all markers in plink-dataset — perMarkerQC"><meta name="description" content="perMarkerQC checks the markers in the plink dataset for their missingness
rates across samples, their deviation from Hardy-Weinberg-Equilibrium (HWE)
and their minor allele frequencies (MAF). Per default, it assumes that IDs of
individuals that have failed perIndividualQC have been written
to qcdir/name.fail.IDs and removes these individuals when computing
missingness rates, HWE p-values and MAF. If the qcdir/name.fail.IDs file does
not exist, a message is written to stdout but the analyses will continue for
all samples in the name.fam/name.bed/name.bim dataset.
Depicts i) SNP missingness rates (stratified by minor allele
frequency) as histograms, ii) p-values of HWE exact test (stratified by all
and low p-values) as histograms and iii) the minor allele frequency
distribution as a histogram."><meta property="og:description" content="perMarkerQC checks the markers in the plink dataset for their missingness
rates across samples, their deviation from Hardy-Weinberg-Equilibrium (HWE)
and their minor allele frequencies (MAF). Per default, it assumes that IDs of
individuals that have failed perIndividualQC have been written
to qcdir/name.fail.IDs and removes these individuals when computing
missingness rates, HWE p-values and MAF. If the qcdir/name.fail.IDs file does
not exist, a message is written to stdout but the analyses will continue for
all samples in the name.fam/name.bed/name.bim dataset.
Depicts i) SNP missingness rates (stratified by minor allele
frequency) as histograms, ii) p-values of HWE exact test (stratified by all
and low p-values) as histograms and iii) the minor allele frequency
distribution as a histogram."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="standard" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">plinkQC</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html" aria-label="ho"><span class="fa fa-home"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-documentation" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Documentation"><span class="fa fa-file"></span> Documentation</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-documentation"><li><a class="dropdown-item" href="../articles/plinkQC.html">Step-by-step QC</a></li>
    <li><a class="dropdown-item" href="../articles/AncestryCheck.html">Population structure identification by Random Forest</a></li>
    <li><a class="dropdown-item" href="../articles/HapMap.html">HapMap III reference</a></li>
    <li><a class="dropdown-item" href="../articles/Genomes1000.html">1000 Genomes reference</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="nav-link" href="../news/index.html"><span class="fa fa-newspaper-o"></span> news</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html"><span class="fa fa-code"></span> functions</a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/meyer-lab-cshl/plinkQC"><span class="fa fa-github fa-lg"></span> github</a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Quality control for all markers in plink-dataset</h1>
      <small class="dont-index">Source: <a href="https://github.com/meyer-lab-cshl/plinkQC/blob/HEAD/R/markerQC.R" class="external-link"><code>R/markerQC.R</code></a></small>
      <div class="d-none name"><code>perMarkerQC.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>perMarkerQC checks the markers in the plink dataset for their missingness
rates across samples, their deviation from Hardy-Weinberg-Equilibrium (HWE)
and their minor allele frequencies (MAF). Per default, it assumes that IDs of
individuals that have failed <code><a href="perIndividualQC.html">perIndividualQC</a></code> have been written
to qcdir/name.fail.IDs and removes these individuals when computing
missingness rates, HWE p-values and MAF. If the qcdir/name.fail.IDs file does
not exist, a message is written to stdout but the analyses will continue for
all samples in the name.fam/name.bed/name.bim dataset.
Depicts i) SNP missingness rates (stratified by minor allele
frequency) as histograms, ii) p-values of HWE exact test (stratified by all
and low p-values) as histograms and iii) the minor allele frequency
distribution as a histogram.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">perMarkerQC</span><span class="op">(</span></span>
<span>  <span class="va">indir</span>,</span>
<span>  qcdir <span class="op">=</span> <span class="va">indir</span>,</span>
<span>  <span class="va">name</span>,</span>
<span>  do.check_snp_missingness <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  lmissTh <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>  do.check_hwe <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  hweTh <span class="op">=</span> <span class="fl">1e-05</span>,</span>
<span>  do.check_maf <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  macTh <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  mafTh <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  interactive <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  keep_individuals <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  remove_individuals <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  exclude_markers <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  extract_markers <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  legend_text_size <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  legend_title_size <span class="op">=</span> <span class="fl">7</span>,</span>
<span>  axis_text_size <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  axis_title_size <span class="op">=</span> <span class="fl">7</span>,</span>
<span>  title_size <span class="op">=</span> <span class="fl">9</span>,</span>
<span>  subplot_label_size <span class="op">=</span> <span class="fl">9</span>,</span>
<span>  path2plink <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  showPlinkOutput <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-indir">indir<a class="anchor" aria-label="anchor" href="#arg-indir"></a></dt>
<dd><p>[character] /path/to/directory containing the basic PLINK data
files name.bim, name.bed, name.fam files.</p></dd>


<dt id="arg-qcdir">qcdir<a class="anchor" aria-label="anchor" href="#arg-qcdir"></a></dt>
<dd><p>[character] /path/to/directory where results will be written to.
If <code><a href="perIndividualQC.html">perIndividualQC</a></code> was conducted, this directory should be the
same as qcdir specified in <code><a href="perIndividualQC.html">perIndividualQC</a></code>, i.e. it contains
name.fail.IDs with IIDs of individuals that failed QC. User needs writing
permission to qcdir. Per default, qcdir=indir.</p></dd>


<dt id="arg-name">name<a class="anchor" aria-label="anchor" href="#arg-name"></a></dt>
<dd><p>[character] Prefix of PLINK files, i.e. name.bed, name.bim,
name.fam.</p></dd>


<dt id="arg-do-check-snp-missingness">do.check_snp_missingness<a class="anchor" aria-label="anchor" href="#arg-do-check-snp-missingness"></a></dt>
<dd><p>[logical] If TRUE, run
<code><a href="check_snp_missingness.html">check_snp_missingness</a></code>.</p></dd>


<dt id="arg-lmissth">lmissTh<a class="anchor" aria-label="anchor" href="#arg-lmissth"></a></dt>
<dd><p>[double] Threshold for acceptable variant missing rate across
samples.</p></dd>


<dt id="arg-do-check-hwe">do.check_hwe<a class="anchor" aria-label="anchor" href="#arg-do-check-hwe"></a></dt>
<dd><p>[logical] If TRUE, run <code><a href="check_hwe.html">check_hwe</a></code>.</p></dd>


<dt id="arg-hweth">hweTh<a class="anchor" aria-label="anchor" href="#arg-hweth"></a></dt>
<dd><p>[double] Significance threshold for deviation from HWE.</p></dd>


<dt id="arg-do-check-maf">do.check_maf<a class="anchor" aria-label="anchor" href="#arg-do-check-maf"></a></dt>
<dd><p>[logical] If TRUE, run <code><a href="check_maf.html">check_maf</a></code>.</p></dd>


<dt id="arg-macth">macTh<a class="anchor" aria-label="anchor" href="#arg-macth"></a></dt>
<dd><p>[double] Threshold for minor allele cut cut-off, if both mafTh
and macTh are specified, macTh is used (macTh = mafTh\*2\*NrSamples).</p></dd>


<dt id="arg-mafth">mafTh<a class="anchor" aria-label="anchor" href="#arg-mafth"></a></dt>
<dd><p>[double] Threshold for minor allele frequency cut-off.</p></dd>


<dt id="arg-interactive">interactive<a class="anchor" aria-label="anchor" href="#arg-interactive"></a></dt>
<dd><p>[logical] Should plots be shown interactively? When
choosing this option, make sure you have X-forwarding/graphical interface
available for interactive plotting. Alternatively, set interactive=FALSE and
save the returned plot object (p_marker) via ggplot2::ggsave(p=p_marker,
other_arguments) or pdf(outfile) print(p_marker) dev.off().</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>[logical] If TRUE, progress info is printed to standard out.</p></dd>


<dt id="arg-keep-individuals">keep_individuals<a class="anchor" aria-label="anchor" href="#arg-keep-individuals"></a></dt>
<dd><p>[character] Path to file with individuals to be
retained in the analysis. The file has to be a space/tab-delimited text file
with family IDs in the first column and within-family IDs in the second
column. All samples not listed in this file will be removed from the current
analysis. See <a href="https://www.cog-genomics.org/plink/1.9/filter#indiv" class="external-link">https://www.cog-genomics.org/plink/1.9/filter#indiv</a>.
Default: NULL, i.e. no filtering on individuals.</p></dd>


<dt id="arg-remove-individuals">remove_individuals<a class="anchor" aria-label="anchor" href="#arg-remove-individuals"></a></dt>
<dd><p>[character] Path to file with individuals to be
removed from the analysis. The file has to be a space/tab-delimited text file
with family IDs in the first column and within-family IDs in the second
column. All samples listed in this file will be removed from the current
analysis. See <a href="https://www.cog-genomics.org/plink/1.9/filter#indiv" class="external-link">https://www.cog-genomics.org/plink/1.9/filter#indiv</a>.
Default: NULL, i.e. no filtering on individuals.</p></dd>


<dt id="arg-exclude-markers">exclude_markers<a class="anchor" aria-label="anchor" href="#arg-exclude-markers"></a></dt>
<dd><p>[character] Path to file with makers to be
removed from the analysis. The file has to be a text file with a list of
variant IDs (usually one per line, but it's okay for them to just be
separated by spaces). All listed variants will be removed from the current
analysis. See <a href="https://www.cog-genomics.org/plink/1.9/filter#snp" class="external-link">https://www.cog-genomics.org/plink/1.9/filter#snp</a>.
Default: NULL, i.e. no filtering on markers.</p></dd>


<dt id="arg-extract-markers">extract_markers<a class="anchor" aria-label="anchor" href="#arg-extract-markers"></a></dt>
<dd><p>[character] Path to file with makers to be
included in the analysis. The file has to be a text file with a list of
variant IDs (usually one per line, but it's okay for them to just be
separated by spaces). All unlisted variants will be removed from the current
analysis. See <a href="https://www.cog-genomics.org/plink/1.9/filter#snp" class="external-link">https://www.cog-genomics.org/plink/1.9/filter#snp</a>.
Default: NULL, i.e. no filtering on markers.</p></dd>


<dt id="arg-legend-text-size">legend_text_size<a class="anchor" aria-label="anchor" href="#arg-legend-text-size"></a></dt>
<dd><p>[integer] Size for legend text.</p></dd>


<dt id="arg-legend-title-size">legend_title_size<a class="anchor" aria-label="anchor" href="#arg-legend-title-size"></a></dt>
<dd><p>[integer] Size for legend title.</p></dd>


<dt id="arg-axis-text-size">axis_text_size<a class="anchor" aria-label="anchor" href="#arg-axis-text-size"></a></dt>
<dd><p>[integer] Size for axis text.</p></dd>


<dt id="arg-axis-title-size">axis_title_size<a class="anchor" aria-label="anchor" href="#arg-axis-title-size"></a></dt>
<dd><p>[integer] Size for axis title.</p></dd>


<dt id="arg-title-size">title_size<a class="anchor" aria-label="anchor" href="#arg-title-size"></a></dt>
<dd><p>[integer] Size for plot title.</p></dd>


<dt id="arg-subplot-label-size">subplot_label_size<a class="anchor" aria-label="anchor" href="#arg-subplot-label-size"></a></dt>
<dd><p>[integer] Size of the subplot labeling.</p></dd>


<dt id="arg-path-plink">path2plink<a class="anchor" aria-label="anchor" href="#arg-path-plink"></a></dt>
<dd><p>[character] Absolute path to PLINK executable
(<a href="https://www.cog-genomics.org/plink/1.9/" class="external-link">https://www.cog-genomics.org/plink/1.9/</a>) i.e.
plink should be accessible as path2plink -h. The full name of the executable
should be specified: for windows OS, this means path/plink.exe, for unix
platforms this is path/plink. If not provided, assumed that PATH set-up works
and PLINK will be found by <code><a href="https://jeroen.r-universe.dev/sys/reference/exec.html" class="external-link">exec</a></code>('plink').</p></dd>


<dt id="arg-showplinkoutput">showPlinkOutput<a class="anchor" aria-label="anchor" href="#arg-showplinkoutput"></a></dt>
<dd><p>[logical] If TRUE, plink log and error messages are
printed to standard out.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Named [list] with i) fail_list, a named [list] with 1.
SNP_missingness, containing SNP IDs [vector] failing the missingness
threshold lmissTh, 2. hwe, containing SNP IDs [vector] failing the HWE exact
test threshold hweTh and 3. maf, containing SNPs Ids [vector] failing the MAF
threshold mafTh/MAC threshold macTh and ii) p_markerQC, a ggplot2-object
'containing' a sub-paneled plot with the QC-plots of
<code><a href="check_snp_missingness.html">check_snp_missingness</a></code>, <code><a href="check_hwe.html">check_hwe</a></code> and
<code><a href="check_maf.html">check_maf</a></code>, which can be shown by print(p_markerQC).
List entries contain NULL if that specific check was not chosen.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>perMarkerQC wraps around the marker QC functions
<code><a href="check_snp_missingness.html">check_snp_missingness</a></code>, <code><a href="check_hwe.html">check_hwe</a></code> and
<code><a href="check_maf.html">check_maf</a></code>. For details on the parameters and outputs, check
these function documentations.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">indir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package<span class="op">=</span><span class="st">"plinkQC"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">qcdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">name</span> <span class="op">&lt;-</span> <span class="st">"data"</span></span></span>
<span class="r-in"><span><span class="va">path2plink</span> <span class="op">&lt;-</span> <span class="st">'/path/to/plink'</span></span></span>
<span class="r-in"><span><span class="co"># the following code is not run on package build, as the path2plink on the</span></span></span>
<span class="r-in"><span><span class="co"># user system is not known.</span></span></span>
<span class="r-in"><span><span class="co"># All quality control checks</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="co"># run on all markers and individuals</span></span></span>
<span class="r-in"><span><span class="va">fail_markers</span> <span class="op">&lt;-</span> <span class="fu">perMarkerQC</span><span class="op">(</span>indir<span class="op">=</span><span class="va">indir</span>, qcdir<span class="op">=</span><span class="va">qcdir</span>, name<span class="op">=</span><span class="va">name</span>,</span></span>
<span class="r-in"><span>interactive<span class="op">=</span><span class="cn">FALSE</span>, verbose<span class="op">=</span><span class="cn">TRUE</span>, path2plink<span class="op">=</span><span class="va">path2plink</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># run on subset of individuals and markers</span></span></span>
<span class="r-in"><span><span class="va">keep_individuals_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"keep_individuals"</span>,</span></span>
<span class="r-in"><span>package<span class="op">=</span><span class="st">"plinkQC"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">extract_markers_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"extract_markers"</span>,</span></span>
<span class="r-in"><span>package<span class="op">=</span><span class="st">"plinkQC"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">fail_markers</span> <span class="op">&lt;-</span> <span class="fu">perMarkerQC</span><span class="op">(</span>qcdir<span class="op">=</span><span class="va">qcdir</span>, indir<span class="op">=</span><span class="va">indir</span>,</span></span>
<span class="r-in"><span>name<span class="op">=</span><span class="va">name</span>, interactive<span class="op">=</span><span class="cn">FALSE</span>, verbose<span class="op">=</span><span class="cn">TRUE</span>, path2plink<span class="op">=</span><span class="va">path2plink</span>,</span></span>
<span class="r-in"><span>keep_individuals<span class="op">=</span><span class="va">keep_individuals_file</span>, extract_markers<span class="op">=</span><span class="va">extract_markers_file</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Hannah Meyer.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

