<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>plinkQC: Genotype Quality Control with 'PLINK' — plinkQC-package • plinkQC</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="plinkQC: Genotype Quality Control with 'PLINK' — plinkQC-package"><meta name="description" content="Genotyping arrays enable the direct measurement of an individuals genotype at thousands of markers. 'plinkQC' facilitates genotype quality control for genetic association studies as described by Anderson and colleagues (2010) doi:10.1038/nprot.2010.116
. It makes 'PLINK' basic statistics (e.g. missing genotyping rates per individual, allele frequencies per genetic marker) and relationship functions accessible from 'R' and generates a per-individual and per-marker quality control report. Individuals and markers that fail the quality control can subsequently be removed to generate a new, clean dataset. Removal of individuals based on relationship status is optimised to retain as many individuals as possible in the study. Additionally, there is a pre-trained classifier to predict genomic ancestry of human samples."><meta property="og:description" content="Genotyping arrays enable the direct measurement of an individuals genotype at thousands of markers. 'plinkQC' facilitates genotype quality control for genetic association studies as described by Anderson and colleagues (2010) doi:10.1038/nprot.2010.116
. It makes 'PLINK' basic statistics (e.g. missing genotyping rates per individual, allele frequencies per genetic marker) and relationship functions accessible from 'R' and generates a per-individual and per-marker quality control report. Individuals and markers that fail the quality control can subsequently be removed to generate a new, clean dataset. Removal of individuals based on relationship status is optimised to retain as many individuals as possible in the study. Additionally, there is a pre-trained classifier to predict genomic ancestry of human samples."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="standard" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">plinkQC</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html" aria-label="ho"><span class="fa fa-home"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-documentation" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Documentation"><span class="fa fa-file"></span> Documentation</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-documentation"><li><a class="dropdown-item" href="../articles/plinkQC.html">Step-by-step QC</a></li>
    <li><a class="dropdown-item" href="../articles/AncestryCheck.html">Ancestry</a></li>
    <li><a class="dropdown-item" href="../articles/HapMap.html">HapMap III reference</a></li>
    <li><a class="dropdown-item" href="../articles/Genomes1000.html">1000 Genomes reference</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="nav-link" href="../news/index.html"><span class="fa fa-newspaper-o"></span> news</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html"><span class="fa fa-code"></span> functions</a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/meyer-lab-cshl/plinkQC"><span class="fa fa-github fa-lg"></span> github</a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>plinkQC: Genotype Quality Control with 'PLINK'</h1>
      <small class="dont-index">Source: <a href="https://github.com/meyer-lab-cshl/plinkQC/blob/HEAD/R/plinkQC-package.R" class="external-link"><code>R/plinkQC-package.R</code></a></small>
      <div class="d-none name"><code>plinkQC-package.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Genotyping arrays enable the direct measurement of an individuals genotype at thousands of markers. 'plinkQC' facilitates genotype quality control for genetic association studies as described by Anderson and colleagues (2010) <a href="https://doi.org/10.1038/nprot.2010.116" class="external-link">doi:10.1038/nprot.2010.116</a>
. It makes 'PLINK' basic statistics (e.g. missing genotyping rates per individual, allele frequencies per genetic marker) and relationship functions accessible from 'R' and generates a per-individual and per-marker quality control report. Individuals and markers that fail the quality control can subsequently be removed to generate a new, clean dataset. Removal of individuals based on relationship status is optimised to retain as many individuals as possible in the study. Additionally, there is a pre-trained classifier to predict genomic ancestry of human samples.</p>
    </div>


    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p>Useful links:</p><ul><li><p><a href="https://meyer-lab-cshl.github.io/plinkQC/">https://meyer-lab-cshl.github.io/plinkQC/</a></p></li>
<li><p>Report bugs at <a href="https://github.com/meyer-lab-cshl/plinkQC/issues" class="external-link">https://github.com/meyer-lab-cshl/plinkQC/issues</a></p></li>
</ul></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p><strong>Maintainer</strong>: Hannah Meyer <a href="mailto:hannah.v.meyer@gmail.com">hannah.v.meyer@gmail.com</a> (<a href="https://orcid.org/0000-0003-4564-0899" class="external-link">ORCID</a>)</p>
<p>Other contributors:</p><ul><li><p>Caroline Walter <a href="mailto:carowalter26@hotmail.com">carowalter26@hotmail.com</a> [contributor]</p></li>
<li><p>Maha Syed <a href="mailto:maha.s.syed@gmail.com">maha.s.syed@gmail.com</a> [contributor]</p></li>
</ul></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Hannah Meyer.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

