<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Running functions to format data for ancestry prediction — run_ancestry_format • plinkQC</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet"><meta property="og:title" content="Running functions to format data for ancestry prediction — run_ancestry_format"><meta name="description" content="This function runs convert_to_plink2 and rename_variant_identifiers to format
the data for the ancestry identification with superpop_classification"><meta property="og:description" content="This function runs convert_to_plink2 and rename_variant_identifiers to format
the data for the ancestry identification with superpop_classification"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="standard" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">plinkQC</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html" aria-label="ho"><span class="fa fa-home"></span></a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-documentation" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true" aria-label="Documentation"><span class="fa fa-file"></span> Documentation</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-documentation"><li><a class="dropdown-item" href="../articles/plinkQC.html">Step-by-step QC</a></li>
    <li><a class="dropdown-item" href="../articles/AncestryCheck.html">Population structure identification by Random Forest</a></li>
    <li><a class="dropdown-item" href="../articles/HapMap.html">HapMap III reference</a></li>
    <li><a class="dropdown-item" href="../articles/Genomes1000.html">1000 Genomes reference</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="nav-link" href="../news/index.html"><span class="fa fa-newspaper-o"></span> news</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html"><span class="fa fa-code"></span> functions</a></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/meyer-lab-cshl/plinkQC"><span class="fa fa-github fa-lg"></span> github</a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Running functions to format data for ancestry prediction</h1>
      <small class="dont-index">Source: <a href="https://github.com/meyer-lab-cshl/plinkQC/blob/HEAD/R/ancestry.R" class="external-link"><code>R/ancestry.R</code></a></small>
      <div class="d-none name"><code>run_ancestry_format.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function runs convert_to_plink2 and rename_variant_identifiers to format
the data for the ancestry identification with superpop_classification</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">run_ancestry_format</span><span class="op">(</span></span>
<span>  <span class="va">indir</span>,</span>
<span>  <span class="va">name</span>,</span>
<span>  qcdir <span class="op">=</span> <span class="va">indir</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  path2plink2 <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  keep_individuals <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  remove_individuals <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  exclude_markers <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  extract_markers <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  showPlinkOutput <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  format <span class="op">=</span> <span class="st">"@:#[hg38]"</span>,</span>
<span>  plink2format <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  var_format <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">path2load_mat</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-indir">indir<a class="anchor" aria-label="anchor" href="#arg-indir"></a></dt>
<dd><p>[character] /path/to/directory containing the basic PLINK 1.9 data
file name.bim, name.fam, name.bed</p></dd>


<dt id="arg-name">name<a class="anchor" aria-label="anchor" href="#arg-name"></a></dt>
<dd><p>[character] Prefix of PLINK 1.9 files, i.e. name.bim, name.fam,
name.bed</p></dd>


<dt id="arg-qcdir">qcdir<a class="anchor" aria-label="anchor" href="#arg-qcdir"></a></dt>
<dd><p>[character] /path/to/directory where the plink2 data formations
as returned by plink2 –make-pgen will be saved to. User needs writing
permission to qcdir. Per default is qcdir=indir.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>[logical] If TRUE, progress info is printed to standard out.</p></dd>


<dt id="arg-path-plink-">path2plink2<a class="anchor" aria-label="anchor" href="#arg-path-plink-"></a></dt>
<dd><p>[character] Absolute path to PLINK executable
(<a href="https://www.cog-genomics.org/plink/2.0/" class="external-link">https://www.cog-genomics.org/plink/2.0/</a>) i.e.
plink 2 should be accessible as path2plink -h. The full name of the executable
should be specified: for windows OS, this means path/plink.exe, for unix
platforms this is path/plink. If not provided, assumed that PATH set-up works
and PLINK will be found by <code><a href="https://jeroen.r-universe.dev/sys/reference/exec.html" class="external-link">exec</a></code>('plink').</p></dd>


<dt id="arg-keep-individuals">keep_individuals<a class="anchor" aria-label="anchor" href="#arg-keep-individuals"></a></dt>
<dd><p>[character] Path to file with individuals to be
retained in the analysis. The file has to be a space/tab-delimited text file
with family IDs in the first column and within-family IDs in the second
column. All samples not listed in this file will be removed from the current
analysis. See <a href="https://www.cog-genomics.org/plink/1.9/filter#indiv" class="external-link">https://www.cog-genomics.org/plink/1.9/filter#indiv</a>.
Default: NULL, i.e. no filtering on individuals.</p></dd>


<dt id="arg-remove-individuals">remove_individuals<a class="anchor" aria-label="anchor" href="#arg-remove-individuals"></a></dt>
<dd><p>[character] Path to file with individuals to be
removed from the analysis. The file has to be a space/tab-delimited text file
with family IDs in the first column and within-family IDs in the second
column. All samples listed in this file will be removed from the current
analysis. See <a href="https://www.cog-genomics.org/plink/1.9/filter#indiv" class="external-link">https://www.cog-genomics.org/plink/1.9/filter#indiv</a>.
Default: NULL, i.e. no filtering on individuals.</p></dd>


<dt id="arg-exclude-markers">exclude_markers<a class="anchor" aria-label="anchor" href="#arg-exclude-markers"></a></dt>
<dd><p>[character] Path to file with makers to be
removed from the analysis. The file has to be a text file with a list of
variant IDs (usually one per line, but it's okay for them to just be
separated by spaces). All listed variants will be removed from the current
analysis. See <a href="https://www.cog-genomics.org/plink/1.9/filter#snp" class="external-link">https://www.cog-genomics.org/plink/1.9/filter#snp</a>.
Default: NULL, i.e. no filtering on markers.</p></dd>


<dt id="arg-extract-markers">extract_markers<a class="anchor" aria-label="anchor" href="#arg-extract-markers"></a></dt>
<dd><p>[character] Path to file with makers to be
included in the analysis. The file has to be a text file with a list of
variant IDs (usually one per line, but it's okay for them to just be
separated by spaces). All unlisted variants will be removed from the current
analysis. See <a href="https://www.cog-genomics.org/plink/1.9/filter#snp" class="external-link">https://www.cog-genomics.org/plink/1.9/filter#snp</a>.
Default: NULL, i.e. no filtering on markers.</p></dd>


<dt id="arg-showplinkoutput">showPlinkOutput<a class="anchor" aria-label="anchor" href="#arg-showplinkoutput"></a></dt>
<dd><p>[logical] If TRUE, plink log and error messages are
printed to standard out.</p></dd>


<dt id="arg-format">format<a class="anchor" aria-label="anchor" href="#arg-format"></a></dt>
<dd><p>[character] This gives the template to rewrite the variant identifier.
A '@' represents the chromosome code, and a '#' represents the base-pair position.</p></dd>


<dt id="arg-plink-format">plink2format<a class="anchor" aria-label="anchor" href="#arg-plink-format"></a></dt>
<dd><p>[logical] If TRUE, data is in plink2 format already and
convert_to_plink2 will not be run</p></dd>


<dt id="arg-var-format">var_format<a class="anchor" aria-label="anchor" href="#arg-var-format"></a></dt>
<dd><p>[logical] If TRUE, variant identifiers are in correct
format already and rename_variant_identifiers will not be run</p></dd>


<dt id="arg-path-load-mat">path2load_mat<a class="anchor" aria-label="anchor" href="#arg-path-load-mat"></a></dt>
<dd><p>[character] /path/to/directory where loading matrices are
kept. This can be downloaded from the github repo. Note that the name of the file
before the .eigenvec.allele or .acount must be included in file path.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Name of file with correct format</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">indir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package<span class="op">=</span><span class="st">"plinkQC"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">qcdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">name</span> <span class="op">&lt;-</span> <span class="st">"data"</span></span></span>
<span class="r-in"><span><span class="va">path2plink</span> <span class="op">&lt;-</span> <span class="st">'/path/to/plink'</span></span></span>
<span class="r-in"><span><span class="co"># the following code is not run on package build, as the path2plink on the</span></span></span>
<span class="r-in"><span><span class="co"># user system is not known.</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="fu">run_ancestry_format</span><span class="op">(</span>indir<span class="op">=</span><span class="va">indir</span>, qcdir<span class="op">=</span><span class="va">qcdir</span>, </span></span>
<span class="r-in"><span>  name<span class="op">=</span><span class="va">name</span>, path2plink2 <span class="op">=</span> <span class="va">path2plink2</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Hannah Meyer.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

